---
title: "<center><h1>eQTL Boxplots Demo</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
---


  *  The code for the __eQTL_boxplots__ function can be found [here](https://github.com/RajLabMSSM/Fine_Mapping/blob/master/echolocatoR/R/eQTL_boxplots.R).

  * The example data is monocyte eQTL data from Fairfax et al. (2014), ['__Innate Immune Activity Conditions the Effect of Regulatory Variants upon Monocyte Gene Expression__'](https://science.sciencemag.org/content/343/6175/1246949.long). 
  * In this study, there is data from four different stimulation conditions. However this __eQTL_boxplots__ can be used for datasets with any other kinds of "conditions"(e.g. tissue type, cell line...) or just a single condition. 


# Import Libraries

```{r Import eQTL_boxplots Function, message=F}
# Import libraries 
library(data.table)
library(dplyr)
library(ggplot2)
library(plotly)

# Import eQTL_boxplots function
source("echolocatoR/R/eQTL_boxplots.R")
```


# Supply Your Arguments

  * For clarity, I've written out what each of the parameters are in separate sections. At the end, we will feed all of these arguments into the __eQTL_boxplots__ function and run it.

## 1. snp_list

  * Supply a list of SNPs (using thier rsIDs) that you want to plot.

```{r snp_list}
snp_list <- c("rs7294519", "rs11175620", "rs76904798")
```

## 2. eQTL_SS_paths

  * Save the root directory as a variable so you don't have to write it out each time.
  * Supply list of paths to each of the summary statistics (SS) files you wish to use (one per condition).
 
```{r eQTL_SS_paths} 
root <- "Data/eQTL/Fairfax_2014"

# Make a list of paths to each of the summary stat files.
eQTL_SS_paths <- file.path(root,
             c("CD14/LRRK2/LRRK2_Fairfax_CD14.txt",
               "IFN/LRRK2/LRRK2_Fairfax_IFN.txt",
               "LPS2/LRRK2/LRRK2_Fairfax_LPS2.txt",
               "LPS24/LRRK2/LRRK2_Fairfax_LPS24.txt"))
print(eQTL_SS_paths)
```

## 3. expression_paths

  * Now we do the same for the gene expression files.
  * Currently, __eQTL_boxplots__ assumes that each expression file is tab-delimited and contains the following columns:
    + [1] __PROBE_ID__ (containing the probe ID, e.g. ILMN_1802380)
    + [2] __1 -> n__ (One column per subject, where the column name correpsonds to the subject's ID. In fairfax, the subject IDs are just numbers (1-421), but these could be any unique identifier)

```{r expression_paths} 
expression_paths <- file.path(root,
              c("CD14/CD14.47231.414.b.txt",
                "IFN/IFN.47231.367.b.txt",
                "LPS2/LPS2.47231.261.b.txt",
                "LPS24/LPS24.47231.322.b.txt"))
print(expression_paths)
```

## 4. probe_path

  * Since eQTL SS and expression files often have the probe listed (instead of the gene symbol it/s targetting ), you'll need to provide a probe mapping file with two columns:
    + [1] __GENE__ (containing gene symbols)
    + [2] __PROBE_ID__ (containing the corresponding probe ID, e.g. ILMN_1802380).
    
  * Often, genes will have multiple probes per gene. By default, __eQTL_boxplots__ only uses the probe with the highest average expression across all samples.

```{r probe_path}
 probe_path <- file.path(root,"gene.ILMN.map")
```

## 5. gene

 * Specify the gene (or list of genes) you want to get the expression data for. __eQTL_boxplots__ will use the probe mapping file to pull out the relevant probe(s) for this gene.
 * In this case, the probe mapping file uses offical HGNC gene symbols, but you should use whatever format your probe mapping file uses (e.g. ENSEMBL IDs).

```{r gene}
gene <- "LRRK2"
```


## 6. genotype_path

  * The genotype file is pretty huge, so you want to work with a subset of it in R. If you already have subset the data to just the SNPs you're interested in...great! Go ahead and use the path to the subsetted file.
  * If you only have the full genotype file, fret not! Simply supply the path to the full genotype data instead AND set __subset_genotype_file = T__. It will automatically search the file for 
    + __NOTE__: Even though this uses grep, this process is still fairly slow and can take several minutes to extract the subset.

```{r genotype_path}
## Using already subsetted file
genotype_path <- file.path(root,"geno.subset.txt")
subset_genotype_file <- F

## Using full genotype file
# genotype_path <- file.path(root,"volunteer_421.impute2.dosage")
# subset_genotype_file <- T
```

## 7. .fam_path

  * Annoyingly, the subject IDs are not included in the genotype file and must be pulled from the corresponding .fam file that was also generated by plink. The genotype subject ID are necessarily in order (1,2,3,4,etc.) so if you assume this your data is going to be all mixed up. 
  * Instead, __eQTL_boxplots__ takes the .fam file and correctly labels the genotype subject IDs for you. 

```{r .fam_path}
.fam_path <- file.path(root,"volunteers_421.fam")
```

## 8. plotting parameters

  * __show_plot__ : You can specify whether you want to actually show the plot ( = TRUE).
  * __SS_annotations__ : You can choose whether to add annotations (Beta, P, FDR) to each boxplot (= TRUE).
  * __interact__ : Make the plot interactive using plotly::ggplotly! (= TRUE). [NOTE: under construction...]

```{r plotting parameters}
show_plot <- T
SS_annotations <- T
interact <- F
```


# Run __eQTL_boxplots__ Function 

  * Time to use your arguments to run __eQTL_boxplots__! In this example, I have supplied 3 SNPs and eQTL data from 4 different conditions, so 12 eQTL boxplots will be generated.
  * If available in the eQTL_SS_paths, __eQTL_boxplots__ will also annotate each boxplot with:
    + Beta (effect size)
    + P (P-value)
    + FDR (False Discovery Rate)

```{r Run eQTL_boxplots Function, fig.width=15, fig.height=12}
merged_eQTL_data <- eQTL_boxplots(
             snp_list = snp_list,
             eQTL_SS_paths = eQTL_SS_paths,
             expression_paths = expression_paths,
             probe_path = probe_path,
             gene = gene,
             genotype_path = genotype_path,
             .fam_path = .fam_path,
             show_plot = show_plot,
             SS_annotations = SS_annotations,
             interact = interact)
```

# Output Data

  * __eQTL_boxplots__ also produces a data.table containing all the data SS, expression, and genotype data used to create the plots.
  * Each row a data point for a given SNP, within a given condition, within a given individual.

```{r Output Data}
createDT(merged_eQTL_data) 
```


