---
title: "<center><h1>Fine Mapping:</h1>Parkinson's Disease</h1></center>" 
# author: 
#     "<div class='container'>
#      <h3>Brian M. Schilder, Bioinformatician II<br>
#      Raj Lab<br>
#      Department of Neuroscience<br>
#      Icahn School of Medicine at Mount Sinai<br>
#      NYC, New York<br>
#      </h3> 
#      <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
#      <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
#      <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
#         <img src='./web/images/brain-icon.png'>
#         <span class='caption'>RAJ LAB</span>
#      <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
#      </div>"
# date: "<br>Most Recent Update:<br> `r Sys.Date()`"
# output: rmarkdown::html_vignette
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: false
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options:
  chunk_output_type: inline
--- 


```{r setup, message=F, warning=F, class.output = "pre"}
# knitr::opts_chunk$set(error = T) # Knit even with errors 
# source("echolocatoR/R/MAIN.R") 
library(dplyr)

library(stringr); library(bsselectR) # devtools::install_github("walkerke/bsselectR")
```

  

# Nalls et. al. (2019) w/ 23andMe

Summarize Fine-mapping Results

## Show all plots  

```{r Gather plots}
# .tabset-dropdown
loci <- list.dirs("./Data/GWAS/Nalls23andMe_2019", recursive = F) %>% basename()
loci <- loci[loci!="_genome_wide"]
plot.paths <- file.path("./Data/GWAS/Nalls23andMe_2019", 
                        loci, "Multi-finemap",
                        paste0(loci,"_ggbio.png"))

# Add link to plot
  # baseURL <- "https://github.com/RajLabMSSM/Fine_Mapping/blob/master/Data/GWAS/Nalls23andMe_2019"
  # links <- data.frame(Link=file.path(baseURL, top_loci_sort$Locus,"Multi-finemap",
  #           paste0(top_loci_sort$Locus,"_ggbio.png")))
  # top_loci_sort <- cbind(links, top_loci_sort)

genes <-  basename(dirname(dirname(plot.paths)))
 "https://rajlabmssm.github.io/Fine_Mapping/Fine_Mapping_PD_results.html"
# file.path("https://rajlabmssm.github.io/Fine_Mapping/Data/Nalls23andMe_2019",
#           genes,"Multi-finemap", paste0(genes,"_ggbio.png"))
# Dropdown

plot.paths <- file.path("https://rajlabmssm.github.io/Fine_Mapping/Data/GWAS/Nalls23andMe_2019",genes,"Multi-finemap",paste0(genes,"_ggbio.png"))
names(plot.paths) <-  genes
 

bsselect(vector = plot.paths, 
         type = "img", 
         selected = "LRRK2", 
         frame_height = "1000", 
         live_search = T, 
         show_tick = T)

# for (p in plot.paths){
#   if(file.exists(p)){
#     try({ 
#     gene <- basename(dirname(dirname(p)))
#     cat('  \n###', gene, '  \n')
#     cat(paste0('![',gene,'](',p,')'))
#     cat('  \n')
#     })
#   }
# }
```


# Session Information

```{r Session Information}
sessionInfo()
print(paste("susieR ", packageVersion("susieR")))
```


  

