---
title: "<center>Fine Mapping:<br>Parkinson's Disease & Alzheimer's Disease</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/'><img src='./web/images/sinai.png'></a>
     </div>"
date: "`r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

```{r setup, message=F, warning=F, dpi = 600}
# devtools::install_local("geneRefineR/", force = T)
# library("echolocatoR") 
source("echolocatoR/R/functions.R") 

library(readxl)
library(DT)
library(data.table)
library(dplyr)
library(ggplot2)
library(plotly)
library(cowplot)
library(ggrepel)
library(curl)
library(biomaRt) 
# Ensembl LD API
library(httr)
library(jsonlite)
library(xml2)
library(gaston)
library(RCurl)
# *** susieR ****
# library(knitrBootstrap) #install_github('jimhester/knitrBootstrap')
library(susieR) # devtools::install_github("stephenslab/susieR")

sessionInfo()
print(paste("susieR ", packageVersion("susieR")))  
# MINERVA PATHS TO SUMMARY STATISTICS DATASETS:
# root <- "../../.."
root <- "~/../../../sc/orga/projects"
Data_dirs = list(
  # ++++++++ GWAS SUMMARY STATS ++++++++ # 
  # Nall et al. (2019) w/ 23andMe
  "Nalls_23andMe" = list(type="Parkinson's GWAS", 
                         topSS="Data/Parkinsons/Nalls_23andMe/Nalls2019_TableS2.xlsx",
                   # fullSS=file.path(root,"pd-omics/data/nallsEtAl2019/23andme/PD_all_post30APRIL2015_5.2_extended.txt")),
                   fullSS=file.path(root,"pd-omics/data/nallsEtAl2019/combined_meta/nallsEtAl2019_allSamples_allVariants.mod.txt"),
                   reference="https://www.biorxiv.org/content/10.1101/388165v3"),  
  
  ## IGAP 
  "Lambert_2013" = list(type="Alzheimer's GWAS", 
                topSS="Data/Alzheimers/Lambert_2013/Lambert_2019_AD_GWAS.xlsx",
                fullSS=file.path(root,"ad-omics/data/AD_GWAS/Lambert_2013/IGAP_stage_1.1000G.phase3.20130502.tsv"),
                reference="https://www.nature.com/articles/ng.2802"),
  
  ## Marioni et al. (2018) 
  "Marioni_2018" = list(type="Alzheimer's GWAS", 
                        topSS="Data/Alzheimers/Marioni_2018/Marioni2018_supplementary_tables.xlsm",
                        fullSS=file.path(root,"ad-omics/data/AD_GWAS/Marioni_2018/Marioni2018.4_UK_Biobank_IGAP_17May2018.1000G.phase3.20130502.tsv"),
                        reference="https://www.nature.com/articles/s41398-018-0150-6"),
  
  ## Jansen et al. (2018) 
  "Posthuma_2018" = list(type="Alzheimer's GWAS", 
                         topSS="Data/Alzheimers/Posthuma_2018/Posthuma2018_suppTables.xlsx", 
                         fullSS=file.path(root,"ad-omics/data/AD_GWAS/Posthuma_2018/phase3.beta.se.hrc.txt"),
                reference="https://www.nature.com/articles/s41588-018-0311-9"),
  
  ## Kunkle et al. (2018) Alzheimer's GWAS 
  "Kunkle_2019" = list(type="Alzheimer's GWAS",
                       topSS="Data/Alzheimers/Kunkle_2019/Kunkle2019_supplementary_tables.xlsx", 
                       fullSS=file.path(root,"ad-omics/data/AD_GWAS/Kunkle_2019/Kunkle_etal_Stage1_results.txt"),
                       # fullSS_stage2=file.path(root,"ad-omics/data/AD_GWAS/Kunkle_etal_Stage2_results.txt"),
                       reference="https://www.nature.com/articles/s41588-019-0358-2"),

  # ++++++++ eQTL SUMMARY STATS ++++++++ #
  ## MESA eQTLs: African Americans
  "MESA_AFA" = list(type="eQTL",
                    topSS="Data/eQTL/MESA/AFA_eQTL_PTK2B.txt",
                    fullSS=file.path(root,"ad-omics/data/mesa/AFA_cis_eqtl_summary_statistics.txt"), 
                    reference="https://www.nhlbi.nih.gov/science/multi-ethnic-study-atherosclerosis-mesa"),
  ## MESA eQTLs: Caucasians
  "MESA_CAU" = list(type="eQTL",
                    topSS="Data/eQTL/MESA/CAU_eQTL_PTK2B.txt",
                    fullSS=file.path(root,"ad-omics/data/mesa/CAU_cis_eqtl_summary_statistics.txt"),
                    reference="https://www.nhlbi.nih.gov/science/multi-ethnic-study-atherosclerosis-mesa"),
  ## MESA eQTLs: Hispanics
  "MESA_HIS" = list(type="eQTL",
                    topSS="Data/eQTL/MESA/HIS_eQTL_PTK2B.txt",
                    fullSS=file.path(root,"ad-omics/data/mesa/HIS_cis_eqtl_summary_statistics.txt"),
                    reference="https://www.nhlbi.nih.gov/science/multi-ethnic-study-atherosclerosis-mesa") 
  )

Data_dirs_table <- Data_dirs_to_table(Data_dirs, writeCSV = "Results/directories_table.csv")

# Direct to local data files if not working on Minerva
if(getwd()=="/Users/schilder/Desktop/Fine_Mapping"){
  vcf_folder = F # Use internet if I'm working from my laptop
} else{
  vcf_folder = F#"../1000_Genomes_VCFs/Phase1" # Use previously downloaded files if working on Minerva node
}  

force_new_subset = F
allResults <- list()
```
 
# Parkinson's Disease

## Nalls et. al. (2019) w/ 23andMe {.tabset .tabset-fade .tabset-pills}

```{r PD: Nalls_23andMe, results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Nalls_23andMe"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats")

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("LRRK2","SNCA","VPS13C"),
                 file_path= Data_dirs[[dataset_name]]$fullSS,#"./LRRK2_EUR_subset.txt",#
                 chrom_col = "CHR", position_col = "POS", snp_col = "RSID",
                 pval_col = "p", effect_col = "beta", stderr_col = "se", 
                 vcf_folder = vcf_folder, superpopulation = "EUR", force_new_subset = force_new_subset) 
#c("LRRK2","SNCA","VPS13C","GCH1"),
```


# Alzheimer's Disease

## Posthuma (2018) {.tabset .tabset-fade .tabset-pills}

```{r AD: Posthuma (2018) , results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Posthuma_2018"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  sheet = "Table S8",
  chrom_col = "Chr", position_col = "bp", snp_col="SNP",
  pval_col="P", effect_col="Zscore", gene_col="nearestGene",
  caption= "Posthuma et al. (2018) AD GWAS Summary Stats")

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),
                 file_path=Data_dirs[[dataset_name]]$fullSS,#"Data/Alzheimers/Posthuma_2018/phase3.beta.se.hrc.txt",#
                 effect_col = "BETA", stderr_col = "SE", position_col = "BP",
                 snp_col = "SNP", pval_col = "P",
                 vcf_folder = vcf_folder, superpopulation = "EUR",
                 minPos=NULL, maxPos=27440000, force_new_subset = force_new_subset)
```

## Marioni (2018) {.tabset .tabset-fade .tabset-pills}

```{r AD: Marioni (2018) , results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Marioni_2018"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  sheet = "Table S9",
  chrom_col = "topSNP_chr", position_col = "topSNP_bp", snp_col="topSNP",
  pval_col="p_GWAS", effect_col="b_GWAS", gene_col="Gene",
  caption= "Marioni et al. (2018) AD GWAS Summary Stats")

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),
                 file_path=Data_dirs[[dataset_name]]$fullSS,#"Data/Alzheimers/Marioni_2018/Marioni2018.4_UK_Biobank_IGAP_17May2018.1000G.phase3.20130502.tsv", #,
                 effect_col = "BETA", stderr_col = "SE", position_col = "POS", 
                 snp_col = "SNP",chrom_col = "CHROM", pval_col = "PVAL",
                 vcf_folder = vcf_folder, superpopulation = "EUR",
                 minPos=NULL, maxPos=27440000, force_new_subset = force_new_subset)
```


## Lambert (2013) (IGAP) {.tabset .tabset-fade .tabset-pills}

Filtering out the CLU portion of the locus prevents susieR from identifying a credible set.
```{r AD: Lambert (2013), results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Lambert_2013"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  sheet = 1,
  chrom_col = "CHR", position_col = "Position", snp_col="SNP",
  pval_col="Overall_P", effect_col="Overall_OR", gene_col="Closest gene",
  caption= "Lambert (2013) (IGAP): AD GWAS Summary Stats")
# genes <- snps_to_genes(snp_list=top_SNPs$SNP )

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"), 
                   file_path=Data_dirs[[dataset_name]]$fullSS,#"Data/Alzheimers/Lambert_2013/IGAP_stage_1.1000G.phase3.20130502.tsv",#
                   chrom_col = "CHROM",  pval_col = "PVAL", snp_col = "SNP",
                   effect_col = "BETA", stderr_col = "SE", position_col = "POS", 
                   vcf_folder = vcf_folder, num_causal = 1, superpopulation = "EUR",
                   minPos=NULL, maxPos=27440000, force_new_subset = force_new_subset)  
```

## Kunkle (2019) (IGAP) {.tabset .tabset-fade .tabset-pills}

Exclude CLU-associated portion of locus by removing all SNPs over position 8:27440000.   
The CLU-gene technically starts at 8:27456253, but its associated SNPs are further upstream.
```{r AD: Kunkle (2019), results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "Kunkle_2019"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  sheet = "Supplementary Table 6",
  chrom_col = "Chr", position_col = "Basepair", snp_col="SNP",
  pval_col="P", effect_col="Beta", gene_col="LOCUS",
  caption= "Kunkle (2019) (IGAP): AD GWAS Summary Stats")
# genes <- snps_to_genes(snp_list=top_SNPs$SNP )

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),  
                   file_path=Data_dirs[[dataset_name]]$fullSS,#"Data/Alzheimers/Kunkle_2019/Kunkle_etal_Stage1_results.txt",#
                   chrom_col = "Chromosome",  position_col = "Position", pval_col = "Pvalue",
                   snp_col = "MarkerName", effect_col = "Beta", stderr_col = "SE", 
                   vcf_folder = vcf_folder, num_causal = 1, superpopulation = "EUR",
                   minPos=NULL, maxPos=27440000, file_sep=" ", force_new_subset = force_new_subset) # Exclude CLU-associated portion of locus
```

# eQTL

- AFR = AFA = African  
- AMR = Ad Mixed American  
- EAS = East Asian  
- EUR = CAU = European  
- SAS = South Asian  
- HIS = Hispanic

## MESA - AFA {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - AFA, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_AFA"
superpopulation <- "AFA"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene =  "PTK2B", 
                                         fullSS_path = Data_dirs[[dataset_name]]$fullSS, 
                                         force_new_subset = force_new_subset)
```

## MESA - CAU {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - CAU, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_CAU"
superpopulation <- "CAU"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene = "PTK2B", 
                                         fullSS_path = Data_dirs[[dataset_name]]$fullSS, 
                                         force_new_subset = force_new_subset)
```

## MESA - HIS {.tabset .tabset-fade .tabset-pills}

Used the Ad Mixed American (AMR) reference panel given the absence of a Hispanic reference panel for 1KG_Phase1.
```{r eQTL: MESA - HIS, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_HIS"
superpopulation <- "HIS"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene = "PTK2B", 
                                       fullSS_path = Data_dirs[[dataset_name]]$fullSS, 
                                       force_new_subset = force_new_subset)
```

# Merge Results

```{r Merge Results}
merged_results <- merge_finemapping_results(allResults, csv_path ="Results/merged_finemapping_results.csv")
createDT(merged_results, caption = "Finemapped Credible Sets for Multiple Datasets")
# table(as.character(merged_results$SNP))
```

