---
title: "<center>Fine Mapping:<br>Parkinson's Disease & Alzheimer's Disease</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     
     <a href='https://github.com/RajLabMSSM'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/'><img src='./web/images/sinai.png'></a>
     </div>"
date: "`r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

```{r setup, message=F, warning=F, dpi = 600}
# devtools::install_local("geneRefineR/", force = T)
# library("geneRefineR") 
source("geneRefineR/R/functions.R") 

library(readxl)
library(DT)
library(data.table)
library(dplyr)
library(ggplot2)
library(plotly)
library(cowplot)
library(ggrepel)
library(curl)
library(biomaRt) 
# Ensembl LD API
library(httr)
library(jsonlite)
library(xml2)
library(gaston)
library(RCurl)
# *** susieR ****
# library(knitrBootstrap) #install_github('jimhester/knitrBootstrap')
library(susieR) # devtools::install_github("stephenslab/susieR")

sessionInfo()
print(paste("susieR ", packageVersion("susieR")))  
# MINERVA PATHS TO SUMMARY STATISTICS DATASETS:
root <- "../../.."
Data_dirs = list(
  # ++++++++ GWAS SUMMARY STATS ++++++++ # 
  # Nall et al. (2019) w/ 23andMe Parkinson's GWAS 
  "Nalls_23andMe" = list(topSS="Data/Parkinsons/Nalls_23andMe/23andMe_indexSNPs.xlsx",
                   # fullSS=file.path(root,"pd-omics/data/nallsEtAl2019/23andme/PD_all_post30APRIL2015_5.2_extended.txt")),
                   fullSS=file.path(root,"pd-omics/data/nallsEtAl2019/combined_meta/nallsEtAl2019_allSamples_allVariants.mod.txt")),  
  ## IGAP Alzheimer's GWAS
  "IGAP" = list(topSS="Data/Alzheimers/IGAP/Lambert_2019_AD_GWAS.xlsx",
                fullSS=file.path(root,"ad-omics/data/AD_GWAS/latest/IGAP_stage_1.1000G.phase3.20130502.tsv")),
  ## Marioni et al. (2018) Alzheimer's GWAS
  "Marioni_2018" = list(topSS=NA,
                        fullSS=file.path(root,"ad-omics/data/AD_GWAS/latest/Marioni2018.4_UK_Biobank_IGAP_17May2018.1000G.phase3.20130502.tsv")),
  ## Posthuma et al. (2018) Alzheimer's GWAS 
  "Posthuma_2018" = list(topSS="Data/Alzheimers/Posthuma/AD_target_SNP.xlsx", 
                fullSS=file.path(root,"ad-omics/data/AD_GWAS/latest/IGAP_stage_1.1000G.phase3.20130502.tsv")),

  # ++++++++ eQTL SUMMARY STATS ++++++++ #
  ## MESA eQTLs: African Americans
  "MESA_AFA" = list(topSS="Data/eQTL/MESA/AFA_eQTL_PTK2B.txt",
                   fullSS=file.path(root,"ad-omics/data/mesa/AFA_cis_eqtl_summary_statistics.txt")),
  ## MESA eQTLs: Caucasians
  "MESA_CAU" = list(topSS="Data/eQTL/MESA/CAU_eQTL_PTK2B.txt",
                   fullSS=file.path(root,"ad-omics/data/mesa/CAU_cis_eqtl_summary_statistics.txt.gz")),
  ## MESA eQTLs: Hispanics
  "MESA_HIS" = list(topSS="Data/eQTL/MESA/HIS_eQTL_PTK2B.txt",
                   fullSS=file.path(root,"ad-omics/data/mesa/HIS_cis_eqtl_summary_statistics.txt.gz")) 
  )
 

# Direct to local data files if not working on Minerva
if(getwd()=="/Users/schilder/Desktop/Fine_Mapping"){
  vcf_folder = F # Use internet if I'm working from my laptop
  Data_dirs$IGAP$fullSS <- "Data/Alzheimers/IGAP/IGAP_stage_1.1000G.phase3.20130502.tsv"
  Data_dirs$Marioni_2018$fullSS <- "Data/Alzheimers/Marioni_2018/Marioni2018.4_UK_Biobank_IGAP_17May2018.1000G.phase3.20130502.tsv"
  Data_dirs$Posthuma_2018$fullSS <- "Data/Alzheimers/Posthuma/phase3.beta.se.hrc.txt"
  Data_dirs$MESA_CAU$fullSS <- "Data/eQTL/MESA/CAU_cis_eqtl_summary_statistics.txt"
} else{
  vcf_folder = "../1000_Genomes_VCFs/Phase1" # Use previously downloaded files if working on Minerva node
}  
```
 
# Parkinson's Disease

## Nalls et. al. (2019) w/ 23andMe {.tabset .tabset-fade .tabset-pills}

```{r PD: Nalls_23andMe, results = 'asis', fig.height=8, fig.width=7} 
top_SNPs <- import_topSNPs(
  file_path = Data_dirs$Nalls_23andMe$topSS,
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats")


finemapped_PD_Nalls_23andMe <- finemap_geneList(top_SNPs, geneList=c("LRRK2"),#c("LRRK2","GBAP1","SNCA","VPS13C","GCH1"),
                 file_path=Data_dirs$Nalls_23andMe$fullSS,
                  snp_col = "MarkerName", pval_col = "P.value", vcf_folder = vcf_folder)
```


# Alzheimer's Disease

## Posthuma (2018) {.tabset .tabset-fade .tabset-pills}

```{r AD: Posthuma (2018) , results = 'asis', fig.height=8, fig.width=7, eval=F} 
top_SNPs <- import_topSNPs(
  file_path = Data_dirs$Posthuma_2018$topSS,
  sheet = 3,
  chrom_col = "Chr", position_col = "bp", snp_col="SNP",
  pval_col="P", effect_col="Z", gene_col="Gene",
  caption= "Posthuma et al. (2018) AD GWAS Summary Stats")


finemapped_AD_Posthuma <- finemap_geneList(top_SNPs, geneList=c("CLU/PTK2B","APOE"),
                 file_path=Data_dirs$Posthuma_2018$fullSS,
                 effect_col = "BETA", stderr_col = "SE", position_col = "BP", 
                 vcf_folder = vcf_folder)
```

## IGAP {.tabset .tabset-fade .tabset-pills}

```{r AD: IGAP, results = 'asis', fig.height=8, fig.width=7} 
top_SNPs <- import_topSNPs(
  file_path = Data_dirs$IGAP$topSS,
  sheet = 1,
  chrom_col = "CHR", position_col = "Position", snp_col="SNP",
  pval_col="Overall_P", effect_col="Overall_OR", gene_col="Closest gene",
  caption= "IGAP: AD GWAS Summary Stats")
# genes <- snps_to_genes(snp_list=top_SNPs$SNP )

finemapped_AD_IGAP <- finemap_geneList(top_SNPs, geneList=c("CLU"), 
                   file_path=Data_dirs$IGAP$fullSS,
                   chrom_col = "CHROM",  pval_col = "PVAL", snp_col = "SNP",
                   effect_col = "BETA", stderr_col = "SE", position_col = "POS", 
                   vcf_folder = vcf_folder, num_causal = 1)
```

# eQTL

AFR = AFA = African
AMR = Ad Mixed American
EAS = East Asian
EUR = CAU = European
SAS = South Asian

## MESA - AFA {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - AFA, results = 'asis', fig.height=8, fig.width=7}
population <- "AFA"
finemapped_eQTL_MESA.AFA <- finemap_eQTL(population, gene =  "PTK2B", 
                                         fullSS_path = paste("Data/eQTL/MESA/",population,
                                                             "_cis_eqtl_summary_statistics.txt",sep=""))
```

## MESA - CAU {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - CAU, results = 'asis', fig.height=8, fig.width=7}
population <- "CAU"
finemapped_eQTL_MESA.CAU <- finemap_eQTL(population, gene = "PTK2B", 
                                         fullSS_path = paste("Data/eQTL/MESA/",population,
                                                             "_cis_eqtl_summary_statistics.txt",sep="")
                                         )
```

## MESA - HIS {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - HIS, results = 'asis', fig.height=8, fig.width=7}
population <- "HIS"
finemapped_eQTL_MESA.HIS <- finemap_eQTL(population, gene = "PTK2B", 
                                         fullSS_path = paste("Data/eQTL/MESA/",population,
                                                             "_cis_eqtl_summary_statistics.txt",sep=""))
```


# Merge Results

```{r Merge Results}
# finemapped_eQTL_MESA.AFA <- finemapped_eQTL_MESA.HIS<- finemapped_eQTL_MESA.CAU
named_results_list <- list("eQTL_MESA.AFA"=finemapped_eQTL_MESA.AFA, 
                           "eQTL_MESA.CAU"=finemapped_eQTL_MESA.CAU, 
                           "eQTL_MESA.HIS"=finemapped_eQTL_MESA.HIS)
merged_results <- merge_finemapping_results(named_results_list, csv_path ="merged_results.csv")
createDT(merged_results, caption = "Finemapped Credible Sets for Multiple Datasets")

table(as.character(merged_results$SNP))
```

