---
title: "<center><h1>Fine Mapping:</h1>Parkinson's Disease</h1></center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 


```{r setup, message=F, warning=F, dpi = 600, class.output = "pre"}
# knitr::opts_chunk$set(error = T) # Knit even with errors
# devtools::install_local("echolocatoR/", force = T)
# library("echolocatoR")
source("echolocatoR/R/MAIN.R")
# MINERVA PATHS TO SUMMARY STATISTICS DATASETS: 
# Data_dirs <- read.csv("./Data/directories_table.csv")
finemap_results <- list()
coloc_results <- list()
```

  

# Nalls et. al. (2019) w/ 23andMe

## Statistical Fine-mapping {.tabset .tabset-fade .tabset-pills} 

Parkinson's Disease GWAS.  

```{r PD: Nalls23andMe_2019, results = 'asis', fig.height=10, fig.width=7}
# fig.show='hold'
dataset_name <- "Nalls23andMe_2019"
top_SNPs <- Nalls_top_SNPs <- import_topSNPs(
  topSS_path = Directory_info(dataset_name, "topSS"),
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats",
  group_by_locus = T, 
  locus_col = "Locus Number",
  remove_variants = "rs34637584")
# Manually add new SNP of interest 
top_SNPs <- data.table::rbindlist(list(top_SNPs, 
                                      data.table::data.table(CHR=14,POS=55360203, SNP="rs3783642", 
                                                   P=0, Effect=1, Gene="ATG14", Locus=NA ),
                                      data.table::data.table(CHR=12,POS=53797236, SNP="rs34656641",
                                                   P=0, Effect=1, Gene="SP1", Locus=NA ),
                                      data.table::data.table(CHR=5,POS=126181862, SNP="rs959573",
                                                   P=0, Effect=1, Gene="LMNB1", Locus=NA ),
                                      data.table::data.table(CHR=17,POS=40609405, SNP="rs9897702",
                                                   P=0, Effect=1, Gene="ATP6V0A1", Locus=NA )
                                      ), use.names = T)
gene_list <- unique(top_SNPs$Gene)

finemap_results[[dataset_name]] <- finemap_gene_list(top_SNPs,
                                               gene_list = unique(c("LRRK2", gene_list)), #
                                               trim_gene_limits = F,#c(T, rep(F,length(gene_list))),
                                               dataset_name = dataset_name,
                                               dataset_type = "GWAS",
                                               query_by ="coordinates",
                                               subset_path = "auto", 
                                               finemap_methods = c("ABF","SUSIE","POLYFUN_SUSIE","FINEMAP"),
                                                #,"ABF","FINEMAP","COJO"), 
                                               force_new_LD = F,    
                                               force_new_subset = T,
                                               force_new_finemap = T, 
                 fullSS_path = Directory_info(dataset_name, "fullSS.local"),
                 chrom_col = "CHR", position_col = "POS", snp_col = "RSID",
                 pval_col = "p", effect_col = "beta", stderr_col = "se", 
                 freq_col = "freq", MAF_col = "calculate",
                 A1_col = "A1",
                 A2_col = "A2",
                 bp_distance = 100000,#500000*2,
                 download_reference = T, 
                 LD_reference = "UKB",#"1KG_Phase1",
                 superpopulation = "EUR", 
                 LD_block = F, 
                 min_MAF = 0,
                 remove_variants = c("rs34637584"),
                 remove_correlates = c("rs34637584"),
                 conditioned_snps = "auto", # Use the lead SNP for each gene  
                 n_causal = 10, 
                 plot_types = c("simple"), #"fancy" 
                 remove_tmps = F)
```

# Summarize Fine-mapping Results

## Gather Annotations

```{r Merge Results}
merged_results <- merge_finemapping_results(minimum_support=1,
                                            include_leadSNPs=T,
                                            xlsx_path="./Data/annotated_finemapping_results.xlsx",
                                            from_storage=T,
                                            haploreg_annotation=T,
                                            biomart_annotation=T, 
                                            verbose = F)
createDT(subset(merged_results, Consensus_SNP == T))
```


# Session Information

```{r Session Information}
sessionInfo()
print(paste("susieR ", packageVersion("susieR")))
```


  

