---
title: "<center><h1>Fine Mapping:</h1>Alzheimer's Disease</h1></center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 
 
```{r setup, message=F, warning=F, dpi = 600} 
source("echolocatoR/R/MAIN.R") 
finemap_results <- list()
```


# PTK2B/CLU

The PTK2B/CLU is a tricky locus with multiple peaks (see Table S6 of Kunkle et al. (2019) for GCTA-COJO results).
We therefore fine-mapped it a little differently:  
  1. Identify the minimum and maximum span of SNPs that are in LD (r2) with the lead common SNP within the LRRK2 gene.
  3. Fine-map each region separately

```{r PTK2B/CLU, eval=F}
quickstart_AD("CLU")
lead.snp <- subset(finemap_DT, leadSNP)
merged <-data.table:::merge.data.table(finemap_DT, 
                              data.table::data.table(r2=LD_matrix[lead.snp$SNP,]^2,
                                                     SNP=names(LD_matrix[lead.snp$SNP,])), 
                              by="SNP") %>% 
  subset(r2>0.1) 
dim(merged)
min_POS <- min(merged$POS)
max_POS <- max(merged$POS)
```

# Kunkle et. al. (2019)

## Statistical Fine-mapping {.tabset .tabset-fade .tabset-pills} 

Parkinson's Disease GWAS.
```{r Kunkle_2019, results = 'asis', fig.height=10, fig.width=7, fig.show='hold'} 
dataset_name <- "Kunkle_2019"
top_SNPs <- import_topSNPs(
              topSS_path = Directory_info(dataset_name, "topSS"),
              sheet = "Supplementary Table 8",
              chrom_col = "CHR", 
              position_col = "POS", 
              snp_col="Top Associated SNV",
              pval_col="P_placeholder", 
              effect_col="Beta_placeholder", 
              gene_col="Lead SNV Gene",
              locus_col = "Lead SNV Gene",
              caption= "Kunkle et al. (2019) GWAS Summary Stats",
              group_by_locus = T)  



finemap_results[[dataset_name]] <- finemap_loci(top_SNPs,
                                               loci =  c("PTK2B"),#,"BIN1","CLU","TREM2"),
                                               trim_gene_limits = F,
                                               dataset_name = dataset_name,
                                               dataset_type = "GWAS",
                                               query_by ="tabix",
                                               subset_path = "auto", 
                                               finemap_method = c("ABF","SUSIE","POLYFUN_SUSIE","FINEMAP"),
                                               force_new_subset = F,
                                               force_new_LD = F,
                                               force_new_finemap = T,  
                 fullSS_path = Directory_info(dataset_name, "fullSS.local"),
                 chrom_col = "Chromosome", 
                 position_col = "Position", 
                 snp_col = "MarkerName",
                 pval_col = "Pvalue", 
                 effect_col = "Beta", 
                 stderr_col = "SE",
                 A1_col = "Effect_allele",
                 A2_col = "Non_Effect_allele",
                 N_cases = 21982,
                 N_controls = 41944, 
                 proportion_cases = "calculate",
                 
                 # min_POS = min_POS,
                 max_POS = min_POS,
                 
                 bp_distance = 500000,
                 download_reference = T, 
                 LD_reference = "UKB",
                 superpopulation = "EUR", 
                 plot_types = "simple",
                 LD_block = F,  
                 min_MAF = 0.001,
                 PP_threshold = .95,
                 n_causal = 5, 
                 remove_tmps = F,
                 server = F)  
```


# Summarise Results

```{r Summarise Results}
merged_DT <- merge_finemapping_results(xlsx_path = "Data/GWAS/Kunkle_2019/Kunkle_2019_results.xlsx",
                                       biomart_annotation = T, dataset = "./Data/GWAS/Kunkle_2019")

```



