---
title: "<center>Fine Mapping:<br>Parkinson's Disease</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     
     <a href='https://github.com/RajLabMSSM'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/'><img src='./web/images/sinai.png'></a>
     </div>"
date: "`r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

# Setup

## Load Libraries

```{r setup, message=F, warning=F,  dpi = 600}
library(readxl)
library(DT)
library(data.table)
library(dplyr)
library(ggplot2)
library(plotly)
library(cowplot)
library(ggrepel)
library(curl)
library(biomaRt)
library(sqldf)
# Ensembl LD API
library(httr)
library(jsonlite)
library(xml2)  
library(gaston)
library(RCurl)
 
# *** susieR ****
# library(knitrBootstrap) #install_github('jimhester/knitrBootstrap')
library(susieR) # devtools::install_github("stephenslab/susieR") 

# *** finemapr ****
## finemapr contains: finemap, CAVIAR, and PAINTOR
# library(finemapr) # devtools::install_github("variani/finemapr")
library(roxygen2) #roxygenize()  

# *** locuscomparer ****
# https://github.com/boxiangliu/locuscomparer
library(locuscomparer); #devtools::install_github("boxiangliu/locuscomparer")

# thm <- knitr::knit_theme$get("bipolar")
# knitr::knit_theme$set(thm)

sessionInfo()
print(paste("susieR ", packageVersion("susieR")))
```
 
## General Functions

```{r}
createDT <- function(DF, caption="", scrollY=400){
  data <- DT::datatable(DF, caption=caption,
    extensions = 'Buttons',
    options = list( dom = 'Bfrtip', 
                    buttons = c('copy', 'csv', 'excel', 'pdf', 'print'), 
                    scrollY = scrollY, scrollX=T, scrollCollapse = T, paging = F,  
                      columnDefs = list(list(className = 'dt-center', targets = "_all"))
    )
  ) 
   return(data)
}
createDT_html <- function(DF, caption="", scrollY=400){
  print( htmltools::tagList( createDT(DF, caption, scrollY)) ) 
} 

# https://stackoverflow.com/questions/42866055/rshiny-download-button-within-rmarkdown
# downloadHandler(filename = function() { 
#     return(paste('Example', input$SS, '.csv', sep=''))
# 
#  }, content = function(file) {
#    write.csv(RandomSample(), file)
#  })
```

# Data Preprocessing

## TSS Data

Use bioMart to get TSS positions for each gene
```{r}
get_TSS_position <- function(gene){ 
  mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
  # att <- listAttributes(mart)
  # grep("transcription", att$name, value=TRUE)
  TSS <- getBM(mart=mart,
               attributes=c("hgnc_symbol","transcription_start_site", "version"), 
               filters="hgnc_symbol", values=gene)
} 
```


## Import GWAS Summary Statistics

For each gene, get the position of top SNP (the one with the greatest effect size/Beta) 
```{r}
## Only the significant subset of results
Nalls_SS_sig <- read_excel("Data/Nalls2018_S2_SummaryStats.xlsx", sheet = "Data")
top_SNPs <- Nalls_SS_sig %>% group_by(`Nearest Gene`) %>% top_n(-1, `P, all studies`) #`Beta, all studies` 
top_SNPs <- cbind(Coord=paste(top_SNPs$CHR,top_SNPs$BP, sep=":"), top_SNPs)
createDT(Nalls_SS_sig, caption = "Nalls (2018): Table S2. Detailed summary statistics on all nominated risk variants, known and novel")
 
# Nall (2018) QTL Stats
# Nalls_QTL <- read_excel("Data/Nalls2018_S4_QTL-MR.xlsx") 
# createDT(Nalls_QTL, caption = "Nalls (2018) Table S4. Expanded summary statistics for QTL Mendelian randomization") 
```

## Get Flanking SNPs

Get all genes surrounding the index SNP (default is 1Mb upstream + 1Mb downstream)
Nalls et al. (2019) data: https://github.com/neurogenetics/meta5
```{r Get Flanking SNPs}    
get_flanking_SNPs <- function(gene, top_SNPs, bp_distance=1000000){ #1000000
  topSNP_sub <- subset(top_SNPs, `Nearest Gene`==gene)
  
  ## Full dataset  
  filePath <- "Data/META.PD.NALLS2014.PRS.tsv"# "Data/nallsEtal2019_no23andMe.tab.txt"
  # Get col names
  # strsplit( readLines(filePath, n = 2), "\t")   
  minPos <- topSNP_sub$BP - bp_distance
  maxPos <- topSNP_sub$BP + bp_distance
  
  geneSubset <- read.csv.sql(filePath, sep="\t", stringsAsFactors=F,
                           sql = paste('select * from file where CHR =',topSNP_sub$CHR,
                                       'AND POS BETWEEN', minPos, 'AND', maxPos)) 
  geneSubset <- geneSubset %>% dplyr::rename(SNP=MarkerName)
  return(geneSubset)
} 
# flankingSNPs <- get_flanking_SNPs("LRRK2", top_SNPs)
```



# susieR 

## Gaston LD

* Nalls et al. 2018: Imputation Panel Notes
    + _"One of the limitations of this study is the use of multiple imputation panels, due to logistic constraints. 
    Adding datasets from non-European populations would be helpful to further improve our granularity in association 
    testing and ability to fine-map loci through integration of more variable LD signatures."_
    + _"Post-Chang 23andMe samples were imputed using a combination of Finch for phasing (an in-house developed fork of Beagle) 
    and miniMac2 for imputation with all-ethnicity samples from the __September 2013 release of 1000 Genomes Phase1__ 
    as reference haplotypes."_
    + _"The Nalls et al . 2014 and Chang et al . 2017 samples were imputed with Minimac2 using 
    __1000 Genomes phase 1 haplotypes__.
    All additional sample series except for the post-Chang et al . 2017 samples from 23andMe were imputed using the 
    __Haplotype Reference Consortium (HRC)__  on the University of Michigan imputation server under default settings 
    with Eagle v2.3 phasing based on reference panel HRC r1.1 2016"_
```{r Gaston LD} 
gaston_LD <- function(gene, flankingSNPs, reference="1KG_Phase1", superpopulation="EUR"){
 
  # Download portion of vcf from 1KG website
  region <- paste(flankingSNPs$CHR[1],":",min(flankingSNPs$POS),"-",max(flankingSNPs$POS), sep="")
  chrom <- flankingSNPs$CHR[1]
  if(reference=="1KG_Phase3"){
    cat("LD Reference Panel = 1KG_Phase3 \n")
    vcf_URL <- paste("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/ALL.chr",chrom,
                     ".phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz",sep="") 
    
      if(!exists("popDat_1KGphase3")){
         popDat <- popDat_1KGphase3 <- read.delim("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel", header = T, row.names = NULL)
         colnames(popDat_1KGphase3) <- c("sample","population","superpopulation","gender")
      } else{popDat <- popDat_1KGphase3}
      
    
  } else if (reference=="1KG_Phase1") {
    cat("LD Reference Panel = 1KG_Phase1 \n")
    vcf_URL <- paste("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/ALL.chr",chrom,
           ".phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf.gz", sep="")
    # Download individual-level metadata
     
      if(!exists("popDat_1KGphase1")){
         popDat <- popDat_1KGphase1  <- read.delim("ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20110521/phase1_integrated_calls.20101123.ALL.panel", header = F, row.names = NULL)
         colnames(popDat) <- c("sample","population","superpopulation","seq_platform1","seq_platform2")
      } else { popDat <- popDat_1KGphase1 }
     
  
  }  
  
  # library(Rsamtools); #BiocManager::install("Rsamtools")
  
  system(paste("tabix -fh ",vcf_URL ,region, "> subset.vcf")) 
  vcf_name <- paste(basename(vcf_URL), ".tbi",sep="") 
  
  # Import w/ gaston and further subset
  bed.file <- read.vcf("subset.vcf")
  ## Subset rsIDs
  bed <- select.snps(bed.file, id %in% flankingSNPs$SNP)
  # Subset Individuals
  selectedInds <- subset(popDat, superpopulation %in% superpopulation) 
  bed <- select.inds(bed, id %in% selectedInds$sample)
  
  # Cleanup extra files
  rm(bed.file)
  file.remove(vcf_name)
  file.remove("subset.vcf")
  
  # Calculate pairwise LD for all SNP combinations
  ld.x <- gaston::LD(bed, lim = c(1,ncol(bed)))  
  # LD plot
  limit <- 20 
  LD.plot( ld.x[1:limit,1:limit], snp.positions = bed@snps$pos[1:limit])
  
  # Double check subsetting
  LD_matrix <- ld.x[row.names(ld.x) %in% flankingSNPs$SNP, colnames(ld.x) %in% flankingSNPs$SNP] 
  return(LD_matrix)
}
# LD_matrix <- gaston_LD("LRRK2",flankingSNPs) 
```

## susieR Function

* Notes on L parameter
    + L is the expected number of causal variants
    + Increasing L increases computational time
    + L=1: Gives a good amount of variation in PIP.
    + L=2: Warns "IBSS algorithm did not converge in 100 iterations!", but gives good variation in PIP.
    + L=3: Warns "IBSS algorithm did not converge in 100 iterations!". All PIPs 1s and 0s.
    + These results seem to be at least partially dependent on whether the ethnic composition of the LD matrix.
* Notes on variance:
    + If 'estimate_residual_variance' = TRUE _without_ providing 'var_y' _and_ L>1, susieR will throw error: 
    __"Estimating residual variance failed: the estimated value is negative"__
    + Running susieR with 'var_y = var(b)' provides _exactly_ the same results. 
* Statistical Terms:
    + posterior inclusion probability (PIP)  
    + coefficient estimate (Beta)
    + Effect allele frequency (EAF)
    + The I^2 statistic describes the percentage of variation across studies that seems not to be due to chance. 
```{r susieR Functions} 
susie_on_gene <- function(gene, top_SNPs, reference="1KG_Phase1"){
  flankingSNPs <- get_flanking_SNPs(gene, top_SNPs)  
  ### Get LD matrix 
  LD_matrix <- gaston_LD(gene, flankingSNPs, reference) 
  ## Turn LD matrix into positive semi-definite matrix
  # LD_matrix2 <- ifelse(matrixcalc::is.positive.semi.definite(LD_matrix), 
  #        LDmatrix,
  #        Matrix::nearPD(LD_matrix)$mat %>% as.matrix() )
  
  ## Subset summary stats to only include SNPs found in query 
  geneSubset <- subset(flankingSNPs, SNP %in% row.names(LD_matrix)) 
  b <- geneSubset$Effect
  se <- geneSubset$StdErr 
  
  # Run Susie 
  fitted_bhat <- susie_bhat(bhat = b, shat = se,
                            R = LD_matrix, 
                            n = nrow(LD_matrix), 
                           
                            L = 1, # 1
                            scaled_prior_variance = 0.1, 
                            estimate_residual_variance = TRUE, # TRUE
                            estimate_prior_variance = FALSE, # FALSE
                            verbose = T,
                            
                            # var_y = var(b),
                            standardize = TRUE
                            ) 
  # Format results
  geneSubset$Coord <- paste(geneSubset$CHR, geneSubset$POS, sep=":")
  susieDF <- data.frame(SNP=names(fitted_bhat$X_column_scale_factors), PIP=fitted_bhat$pip) %>%
  base::merge(subset(geneSubset, select=c("SNP","Effect","P.value", "POS","Coord")), by="SNP") %>% 
    mutate(POS=as.numeric(POS))
  return(susieDF)
}
# susieDF <- susie_on_gene("LRRK2", top_SNPs) 
```

## Plot susieR Results

```{r}
before_after_plots <- function(gene, susieDF, topVariants=3){ 
  roundBreaks <- seq(plyr::round_any(min(susieDF$POS),10000), max(susieDF$POS),500000) 
  ## Before fine-mapping
  geneSubset <- susieDF %>% arrange(desc(abs(Effect)))
  labelSNPs_Effect <- geneSubset[1:topVariants,]
  yLimits <- c(min(-log(susieDF$Effect)), max(-log(susieDF$Effect))+1)
  
  before_plot <- ggplot(geneSubset, aes(x=POS, y=Effect, label=SNP, color= -log(P.value) )) + 
    ylim(yLimits) +
    geom_hline(yintercept=0,alpha=.5, linetype=1, size=.5) +
    geom_point() + 
    geom_point(data=labelSNPs_Effect[1,], pch=21, fill=NA, size=4, colour="red", stroke=1) + 
    geom_segment(aes(xend=POS, yend=0, color= -log(P.value)), alpha=.5) +
    geom_text_repel(data=labelSNPs_Effect, aes(label=SNP), segment.alpha = .2, nudge_y = .05, nudge_x = .5) + 
    labs(title=paste(gene," (",length(susieDF$PIP)," variants)","\nBefore Fine Mapping",sep=""), y="Beta", x="Position",
         color="GWAS\n-log(p-value)") +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_x_continuous(breaks = roundBreaks) 
  
  ## After fine-mapping 
  susieDF <- susieDF %>% arrange(desc(PIP))
  labelSNPs_PIP <- susieDF[1:topVariants,]
  yLimits <- c(min(susieDF$PIP), max(susieDF$PIP)+.1)
  
  after_plot <- ggplot(susieDF, aes(x=POS, y=PIP, label=SNP, color= -log(P.value) )) +  
    ylim(yLimits) +
    geom_hline(yintercept=0,alpha=.5, linetype=1, size=.5) +
    geom_point() +  
    geom_point(data=susieDF[susieDF$PIP == max(susieDF$PIP),], 
               pch=21, fill=NA, size=4, colour="green", stroke=1) +
    geom_segment(aes(xend=POS, yend=yLimits[1], color= -log(P.value))) +
    geom_text_repel(data=labelSNPs_PIP, aes(label=SNP), segment.alpha = .2, nudge_y = .05, nudge_x = .5) + 
    labs(title=paste(gene," (",length(susieDF$PIP)," variants)","\nAfter Fine Mapping",sep=""), y="PIP", x="Position",
         color="GWAS\n-log(p-value)") +
    theme(plot.title = element_text(hjust = 0.5)) +
    scale_x_continuous(breaks = roundBreaks)  
  
  plot_grid(before_plot, after_plot, nrow = 2) %>% print()  
  # susie_plot(fitted_bhat, y="PIP", b=b, add_bar = T)
  
  createDT_html(susieDF, paste("susieR Results: ", gene), scrollY = 200) 
}
# before_after_plots(susieDF)
```

## Report SNP Overlap

```{r Report SNP Overlap}
Nalls_vs_SusieR_overlap <- function(Nalls_SS_sig, susieDF, max_SNPs=10){
    # Get top SNPs from Nalls et all fine mapping
    Nalls_dat  <- subset(Nalls_SS_sig, `Nearest Gene`==gene) %>% 
    arrange(`P, all studies`) %>%
    mutate(SNP=as.character(SNP))
    Nalls_SNPs <-if(max_SNPs > length(Nalls_dat$SNP)){ Nalls_dat$SNP}else{Nalls_dat$SNP[1:max_SNPs]}
  # Get top SNPs from susieR fine mapping
  susieR_dat <- subset(susieDF, PIP!=0) %>% 
    arrange(desc(PIP)) %>% 
    mutate(SNP=as.character(SNP))
  susieR_SNPs <-if(max_SNPs > length(susieR_dat$SNP)){ susieR_dat$SNP}else{susieR_dat$SNP[1:max_SNPs]}
  # Calculate percent
  overlap <-  length(intersect(Nalls_SNPs, susieR_SNPs))
  percentOverlap <- overlap / length(susieR_SNPs) * 100 
  cat("\n",overlap,"/",length(susieR_SNPs), " (",round(percentOverlap,2),
      "%) of SNPs identified by susieR are also identified by Nalls et al. (2018) fine mapping. ","\n",sep="")  
  }
# Nalls_vs_SusieR_overlap(Nalls_SS_sig, susieDF, max_SNPs=10)
```


## Fine Map with Summary Statistics {.tabset .tabset-fade .tabset-pills}

```{r Run susieR, fig.height=8, fig.width=7, results = 'asis'} 
geneList <- c("LRRK2","GBAP1","SNCA","VPS13C","GCH1")#unique(top_SNPs$`Nearest Gene`) 
#Nalls_SS %>% group_by(`Nearest Gene`) %>% tally() %>% subset(n>2)

fineMapped_topSNPs <- data.table()
for (gene in geneList){ 
  cat("\n")
  cat("### ", gene, "\n")
  susieDF <- susie_on_gene(gene, top_SNPs, reference="1KG_Phase1")
  before_after_plots(gene, susieDF)
  Nalls_vs_SusieR_overlap(Nalls_SS_sig, susieDF, max_SNPs=10)
  
  # Create summary table for all genes
  newEntry <- cbind(data.table(Gene=gene), subset(susieDF, PIP==max(PIP)) %>% as.data.table())
  fineMapped_topSNPs <- rbind(fineMapped_topSNPs, newEntry)
  cat("\n")
}
```

## Top SNPs

```{r}
createDT(fineMapped_topSNPs, "Potential Causal SNPs Identified by susieR", scrollY = 200)
```


```{r Individual Data, eval=F, include=F} 
fitted = susie(R, b,
                L = 1,
                estimate_residual_variance = TRUE, 
                estimate_prior_variance = FALSE,
                scaled_prior_variance = 0.1,
                verbose = TRUE)
plot(fitted$pip, fitted_bhat$pip, ylim=c(0,1))
plot(coef(fitted), coef(fitted_bhat)) 
```
 
```{r Using Z-scores , eval=F, include=F}
fitted_z <- susie_z(z_scores, R = R, L = 10)
summary(fitted_z)$cs
susie_plot(fitted_z, y="PIP", b=b)
```


