---
title: "<center><h1>Fine Mapping:</h1>Parkinson's Disease</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

[Scripts to run CAVIAR, FINEMAP (version 1.1) and DAP-G](https://stephenslab.github.io/susie-paper/manuscript_results/pedagogical_example.html)
```{r setup, message=F, warning=F, dpi = 600}
# knitr::opts_chunk$set(error = TRUE) # Knit even with errors
# devtools::install_local("echolocatoR/", force = T)
# library("echolocatoR")
source("echolocatoR/R/MAIN.R")  
sessionInfo()
print(paste("susieR ", packageVersion("susieR")))  
# MINERVA PATHS TO SUMMARY STATISTICS DATASETS: 
Data_dirs <- read.csv("./Data/directories_table.csv")
finemap_results <- list()
coloc_results <- list()
```

 
# Parkinson's Disease

## Nalls et. al. (2019) w/ 23andMe {.tabset .tabset-fade .tabset-pills} 

```{r PD: Nalls23andMe_2019, results = 'asis', fig.height=12, fig.width=9} 
dataset_name <- "Nalls23andMe_2019"

top_SNPs <- Nalls_top_SNPs <- import_topSNPs(
  file_path = Directory_info(Data_dirs, dataset_name, "topSumStats"),
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats")
# Manually add new SNP of interest
top_SNPs <- rbind(top_SNPs, 
                  data.table::data.table(Coord="14:54893485", CHR=14, POS=54893485, SNP="rs3783642", 
                                         P=0, Effect=1, Gene="ATG14"))


gene_list <- c("SNCA", "BST1", "TMEM163", "GPNMB", "CTSB", "ATG14")# "LRRK2_alt","SNCA","VPS13C","C5orf24","IP6K2","KCNIP3","GBF1","SCARB2")#, top_SNPs$Gene) %>% unique()


finemap_results[[dataset_name]] <- finemap_geneList(top_SNPs,
                                               geneList =  c("LRRK2", gene_list),
                                               trim_gene_limits = c(T, rep(F,length(gene_list))),
                                               dataset_name = dataset_name,
                                               dataset_type = "GWAS",
                                               query_by ="coordinates",
                                               subset_path = "auto",
                                               finemap_method = c("SUSIE","ABF","FINEMAP"),# COJO
                                               force_new_LD = F, 
                                               force_new_subset = F,
                                               force_new_finemap = F,
                 # file_path = Directory_info(dataset_name, "fullSumStats"),
                 fullSS_path = "./Data/GWAS/Nalls23andMe_2019/nallsEtAl2019_allSamples_allVariants.mod.txt",
                 chrom_col = "CHR", position_col = "POS", snp_col = "RSID",
                 pval_col = "p", effect_col = "beta", stderr_col = "se", 
                 freq_col = "freq", MAF_col = "calculate",
                 bp_distance = 500000*2,
                 download_reference = T, 
                 LD_reference = "1KG_Phase1",
                 superpopulation = "EUR", 
                 LD_block = F, 
                 min_MAF = 0.01,
                 # remove_variants = c("rs34637584"),
                 remove_correlates = c("rs34637584"),
                 conditioned_snps = c("rs76904798"),
                 n_causal = 5) 
```



```{r Manhattan Plot: Nalls23andMe_2019}
cred_set <- subset(allResults[[dataset_name]], Credible_Set==T & Gene=="LRRK2")

manhattan_plot(# subset_path= get_subset_path(results_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"), gene="LRRK2"),
               subset_path="./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt",
               gene="LRRK2", SNP_list = c("rs76904798","rs34637584","rs117073808"), alt_color_SNPs=c("rs34637584"))
```



# eQTL

- AFR = AFA = African  
- AMR = Ad Mixed American  
- EAS = East Asian  
- EUR = CAU = European  
- SAS = South Asian  
- HIS = Hispanic

## MESA - AFA {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - AFA, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_AFA" 
allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  
                                                superpopulation = "AFA",
                                                dataset_name = dataset_name, 
                                                finemap_method = c("SUSIE","ABF","FINEMAP"),
                                                force_new_LD = F,
                                                force_new_finemap = F, 
                                                force_new_subset = F, 
                                                dataset_type = "eQTL",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "gene",
                   # fullSS_path = Directory_info(dataset_name, "fullSumStats"),
                   fullSS_path = "Data/eQTL/MESA_AFA/LRRK2_AFA_MESA.txt",
                   subset_path = "auto",
                   chrom_col = "chr", position_col = "pos_snps", snp_col="snps",
                   pval_col="pvalue", effect_col="beta", gene_col="gene_name", 
                   stderr_col = "calculate",  tstat_col = "statistic",
                   n_causal = 5,  
                   LD_block = F) 
```

## MESA - CAU {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - CAU, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_CAU"

allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  
                                                superpopulation = "CAU",
                                                dataset_name = dataset_name, 
                                                dataset_type = "eQTL",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "gene",
                   # fullSS_path = Directory_info(dataset_name, "fullSumStats"),
                   fullSS_path = "Data/eQTL/MESA_CAU/LRRK2_CAU_MESA.txt",
                   subset_path = "auto",
                   chrom_col = "chr", position_col = "pos_snps", snp_col="snps",
                   pval_col="pvalue", effect_col="beta", gene_col="gene_name", 
                   stderr_col = "calculate",  tstat_col = "statistic",
                   vcf_folder = vcf_folder, n_causal = 1, 
                   force_new_subset = F,
                   LD_block = F)
```

## MESA - HIS {.tabset .tabset-fade .tabset-pills}

Used the Ad Mixed American (AMR) reference panel given the absence of a Hispanic reference panel for 1KG_Phase1.
```{r eQTL: MESA - HIS, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_HIS" 

allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  
                                                superpopulation = "HIS",
                                                dataset_name = dataset_name, 
                                                dataset_type = "eQTL",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "gene",
                   # fullSS_path = Directory_info(dataset_name, "fullSumStats"),
                   fullSS_path = "Data/eQTL/MESA_HIS/LRRK2_HIS_MESA.txt",
                   subset_path = "auto",
                   chrom_col = "chr", position_col = "pos_snps", snp_col="snps",
                   pval_col="pvalue", effect_col="beta", gene_col="gene_name", 
                   stderr_col = "calculate",  tstat_col = "statistic",
                   vcf_folder = vcf_folder, n_causal = 1, 
                   force_new_subset = F,
                   LD_block = F)
```

### Colocalization 

```{r Colocalization: Nalls23andMe_2019}
# CAU
dataset_name <- "Nalls23andMe_2019.vs.MESA_CAU"
coloc_results[[dataset_name]] <- COLOC(
      dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt", 
      dataset2_path = "./Data/eQTL/MESA_CAU/LRRK2/LRRK2_MESA_CAU_subset.txt",
      dataset1_type = "cc",
      dataset2_type = "quant",
      shared_MAF = 1,
      PP_threshold = 0.8, 
      save_results = T,
      show_plot=T)
# AFA
dataset_name <- "Nalls23andMe_2019.vs.MESA_AFA"
coloc_results[[dataset_name]] <- COLOC(
      dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt", 
      dataset2_path = "./Data/eQTL/MESA_AFA/LRRK2/LRRK2_MESA_AFA_subset.txt",
      shared_MAF = data.table::fread("Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt")$MAF,
      PP_threshold = 0.8, 
      save_results = T,
      show_plot=T,
      plot_subtitle="MESA-AFA eQTL")

# HIS
dataset_name <- "Nalls23andMe_2019.vs.MESA_HIS"
coloc_results[[dataset_name]] <- COLOC(
      dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt", 
      dataset2_path = "./Data/eQTL/MESA_HIS/LRRK2/LRRK2_MESA_HIS_subset.txt",
      shared_MAF = data.table::fread("Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt")$MAF,
      PP_threshold = 0.8, 
      save_results = T,
      show_plot=T,
      plot_subtitle="MESA-HIS eQTL")
```





## Fairfax (2014) {.tabset .tabset-fade .tabset-pills}

+ Use the Nalls et al. (2019) LRRK2 locus to query for SNPs in the eQTL data.  
+ SS Files are SPACE-delimited (not tab-delimited)
```{r Fairfax, results = 'asis', fig.height=8, fig.width=7, eval=F} 

# +++++++++++++++ CD14 Stimulation +++++++++++++++ #
dataset_name <- "Fairfax_2014_CD14"
allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  
                                                superpopulation = "CAU",
                                                dataset_name = dataset_name, 
                                                dataset_type = "eQTL",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "probes",
                                                probe_path = "./Data/eQTL/gene.ILMN.map",
                   file_sep = " ",                            
                   # fullSS_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"),
                   fullSS_path = "./Data/eQTL/Fairfax_2014_CD14/LRRK2_Fairfax_CD14.txt",
                   subset_path = "auto",
                   chrom_col = "CHR", position_col = "POS", snp_col="SNP",
                   pval_col="p-value", effect_col="beta", gene_col="gene", 
                   stderr_col = "calculate",  tstat_col = "t-stat",
                   vcf_folder = vcf_folder, n_causal = 1, 
                   force_new_subset = force_new_subset,
                   LD_block = T)


# +++++++++++++++ IFN Stimulation +++++++++++++++ #
dataset_name <- "Fairfax_2014_IFN"
allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  
                                                superpopulation = "CAU",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "coordinates_merged",
                   file_sep = " ",   
                   dataset_name = dataset_name,
                   dataset_type = "eQTL",
                   # fullSS_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"),
                   fullSS_path = "Data/eQTL/Fairfax_2014_IFN/LRRK2/LRRK2_Fairfax_IFN_mod.txt",
                   subset_path = "auto",
                   chrom_col = "SNP", position_col = "SNP", snp_col="SNP",
                   pval_col="p-value", effect_col="beta", gene_col="gene", 
                   stderr_col = "calculate",  tstat_col = "t-stat",
                   n_causal = 5, 
                   force_new_subset = T,
                   LD_block = T)

# +++++++++++++++ LPS Stimulation (after 2 hours) +++++++++++++++ #
dataset_name <- "Fairfax_2014_LPS2"
allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  superpopulation = "CAU",
                                                top_SNPs = Nalls_top_SNPs, 
                                                # query_by = "probes", # Use if querying the main file
                                                query_by = "coordinates_merged", 
                   file_sep = " ",    
                   dataset_name = dataset_name,
                    dataset_type = "eQTL",
                   # fullSS_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"),
                   fullSS_path = "Data/eQTL/Fairfax_2014_LPS2/LRRK2/LRRK2_Fairfax_LPS2.txt",
                   subset_path = "auto",
                   chrom_col = "SNP", position_col = "SNP", snp_col="SNP",
                   pval_col="p-value", effect_col="beta", gene_col="gene", 
                   stderr_col = "calculate",  tstat_col = "t-stat", 
                   n_causal = 5, 
                   force_new_subset = T,
                   LD_block = T)


# +++++++++++++++ LPS Stimulation (after 24 hours) +++++++++++++++ #
dataset_name <- "Fairfax_2014_LPS24"
allResults[[dataset_name]] <- finemap_geneList( geneList=c("LRRK2"),  superpopulation = "CAU",
                                                top_SNPs = Nalls_top_SNPs, query_by = "coordinates_merged",
                   file_sep = " ",                             
                   fullSS_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"),
                   subset_path = "Data/eQTL/Fairfax/LRRK2_Fairfax_LPS24_subset.txt",
                   chrom_col = "SNP", position_col = "SNP", snp_col="SNP",
                   pval_col="p-value", effect_col="beta", gene_col="gene", 
                   stderr_col = "calculate",  tstat_col = "t-stat",
                   vcf_folder = vcf_folder, n_causal = 1, 
                   force_new_subset = force_new_subset,
                   LD_block = T)
```

### Colocalization

```{r Colocalization: Fairfax_2014}
# CD14 Stimulation
dataset_name <- "Nalls23andMe_2019.vs.Fairfax_2014_CD14"
coloc_results[[dataset_name]] <- COLOC(
      dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt", 
      dataset2_path = "./Data/eQTL/Fairfax_2014_CD14/LRRK2/LRRK2_Fairfax_2014_CD14_subset.txt",
      shared_MAF = data.table::fread("Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt")$MAF,
      plot_subtitle="Fairfax (2014) + CD14 eQTL")

# IFN Stimulation
dataset_name <- "Nalls23andMe_2019.vs.Fairfax_2014_IFN"
coloc_results[[dataset_name]] <- COLOC(
      dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt", 
      dataset2_path = "./Data/eQTL/Fairfax_2014_IFN/LRRK2/LRRK2_Fairfax_2014_IFN_subset.txt",
      shared_MAF = data.table::fread("Data/GWAS/Nalls23andMe_2019/LRRK2/LRRK2_Nalls23andMe_2019_subset.txt")$MAF, 
      plot_subtitle="Fairfax (2014) + IFN eQTL")
```


 

# Merge Results

```{r Merge Results}
merged_results <- merge_finemapping_results(allResults, csv_path ="Data/merged_finemapping_results.csv")
createDT(merged_results, caption = "Finemapped Credible Sets for Multiple Datasets")
# table(as.character(merged_results$SNP))
```

