---
title: "<center><h1>Fine Mapping:</h1>Parkinson's Disease<br>&<br>Alzheimer's Disease</center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

[Scripts to run CAVIAR, FINEMAP (version 1.1) and DAP-G](https://stephenslab.github.io/susie-paper/manuscript_results/pedagogical_example.html)
```{r setup, message=F, warning=F, dpi = 600}
# devtools::install_local("geneRefineR/", force = T)
# library("echolocatoR") 
source("echolocatoR/R/functions.R") 

library(readxl)
library(DT)
library(data.table)
library(dplyr)
library(ggplot2)
library(plotly)
library(cowplot)
library(ggrepel)
library(curl)
library(biomaRt) 
# Ensembl LD API
# library(httr)
# library(jsonlite)
# library(xml2)
# library(gaston)
# library(RCurl) 
# *** susieR ****
# library(knitrBootstrap) #install_github('jimhester/knitrBootstrap')
library(susieR) # devtools::install_github("stephenslab/susieR")
sessionInfo()
print(paste("susieR ", packageVersion("susieR")))  

# MINERVA PATHS TO SUMMARY STATISTICS DATASETS: 
Data_dirs <- read.csv("./Results/directories_table.csv")

# Direct to local data files if not working on Minerva
if(getwd()=="/Users/schilder/Desktop/Fine_Mapping"){
  vcf_folder = F # Use internet if I'm working from my laptop
} else{
  vcf_folder = F#"../1000_Genomes_VCFs/Phase1" # Use previously downloaded files if working on Minerva node
}  

force_new_subset = T
allResults <- list()
```


# CAVIARBF

```{r CAVIARBF, eval=F}  
# system("wget https://bitbucket.org/Wenan/caviarbf/get/7e428645be5e.zip") 
# system("cd CAVIARBF")
# system("make")
# install.packages("echolocatoR/tools/CAVIARBF/caviarbf-r-package/caviarbf_0.2.1.tar.gz", repos = NULL, type = "source") 
library(caviarbf)
# caviarbfFineMapping( )
kunkle <-fread("Data/Alzheimers/Kunkle_2019/PTK2B_Kunkle_2019_subset.txt")
kunkle = subset(kunkle, select = c("MarkerName","Beta","SE"))
system("./caviarbf -z ./example/myfile.Z -r ./example/myfile.LD -t 0 -a 0.1281429 -n 2000 -c 5 -o ./example/myfile.sigma0.1281429.bf")
```

 
# Parkinson's Disease

## Nalls et. al. (2019) w/ 23andMe {.tabset .tabset-fade .tabset-pills}

*"**Rare coding variant burden analysis**:  
We performed kernel-based burden tests on the 113
genes in our PD GWAS loci that contained two or more rare coding variants (MAF< 5% or MAF < 1%). After Bonferroni correction for 113 genes, we identified 7 significant putatively causal genes: 
  - LRRK2, GBA, CATSPER3 **(rs11950533/C5orf24 locus)**
  - LAMB2 **(rs12497850/IP6K2 locus)**
  - LOC442028 **(rs2042477/KCNIP3 locus)**
  - NFKB2 **(rs10748818/GBF1 locus)**, and SCARB2 **(rs6825004 locus)**."*  
-- from [Nalls et al. (2019)](https://www.biorxiv.org/content/10.1101/388165v3.supplementary-material) 
```{r PD: Nalls_23andMe, results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Nalls_23andMe"

top_SNPs <- import_topSNPs(
  file_path = Directory_info(dataset_name, "topSumStats"),
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats")

gene_list <- c("LRRK2")#,"SNCA","VPS13C","C5orf24","IP6K2","KCNIP3","GBF1","SCARB2")#, top_SNPs$Gene) %>% unique()
 
allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=gene_list,
                 file_path = Directory_info(dataset_name, "fullSumStats"),
                 # file_path ="./Data/Parkinsons/Nalls_23andMe/nallsEtAl2019_allSamples_allVariants.mod.txt",
                 chrom_col = "CHR", position_col = "POS", snp_col = "RSID",
                 pval_col = "p", effect_col = "beta", stderr_col = "se", 
                 vcf_folder = vcf_folder, superpopulation = "EUR", force_new_subset = force_new_subset,
                 LD_block = F, min_r2 = 0.2, min_Dprime = F) 
#c("LRRK2","SNCA","VPS13C","GCH1"),
# "GBA","CATSPER3", "LAMB2", "LOC442028", "NFKB2", "SCARB2", "GRN"
```


# Alzheimer's Disease

## Posthuma (2018) {.tabset .tabset-fade .tabset-pills}

```{r AD: Posthuma (2018) , results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Posthuma_2018"

top_SNPs <- import_topSNPs(
  file_path = Data_dirs[[dataset_name]]$topSS,
  sheet = "Table S8",
  chrom_col = "Chr", position_col = "bp", snp_col="SNP",
  pval_col="P", effect_col="Zscore", gene_col="nearestGene",
  caption= "Posthuma et al. (2018) AD GWAS Summary Stats")

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),
                 file_path = Directory_info(dataset_name, "fullSumStats"),
                 #"Data/Alzheimers/Posthuma_2018/phase3.beta.se.hrc.txt",#
                 effect_col = "BETA", stderr_col = "SE", position_col = "BP",
                 snp_col = "SNP", pval_col = "P",
                 vcf_folder = vcf_folder, superpopulation = "EUR",
                 minPos=NULL, maxPos=27440000, force_new_subset = force_new_subset)
```

## Marioni (2018) {.tabset .tabset-fade .tabset-pills}

```{r AD: Marioni (2018) , results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Marioni_2018"

top_SNPs <- import_topSNPs(
  file_path = Directory_info(dataset_name, "topSumStats"),
  sheet = "Table S9",
  chrom_col = "topSNP_chr", position_col = "topSNP_bp", snp_col="topSNP",
  pval_col="p_GWAS", effect_col="b_GWAS", gene_col="Gene",
  caption= "Marioni et al. (2018) AD GWAS Summary Stats")

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),
                 file_path = Directory_info(dataset_name, "fullSumStats"),
                 #"Data/Alzheimers/Marioni_2018/Marioni2018.4_UK_Biobank_IGAP_17May2018.1000G.phase3.20130502.tsv", #,
                 effect_col = "BETA", stderr_col = "SE", position_col = "POS", 
                 snp_col = "SNP",chrom_col = "CHROM", pval_col = "PVAL",
                 vcf_folder = vcf_folder, superpopulation = "EUR",
                 minPos=NULL, maxPos=NULL, force_new_subset = force_new_subset)
```


## Lambert (2013) (IGAP) {.tabset .tabset-fade .tabset-pills}

Filtering out the CLU portion of the locus prevents susieR from identifying a credible set.
```{r AD: Lambert (2013), results = 'asis', fig.height=8, fig.width=7} 
dataset_name <- "Lambert_2013"

top_SNPs <- import_topSNPs(
  file_path = Directory_info(dataset_name, "topSumStats"),
  sheet = 1,
  chrom_col = "CHR", position_col = "Position", snp_col="SNP",
  pval_col="Overall_P", effect_col="Overall_OR", gene_col="Closest gene",
  caption= "Lambert (2013) (IGAP): AD GWAS Summary Stats")
# genes <- snps_to_genes(snp_list=top_SNPs$SNP )

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"), 
                   file_path = Directory_info(dataset_name, "fullSumStats"),
                   # file_path = "Data/Alzheimers/Lambert_2013/IGAP_stage_1.1000G.phase3.20130502.tsv",
                   #Data_dirs[[dataset_name]]$fullSS,#
                   chrom_col = "CHROM",  pval_col = "PVAL", snp_col = "SNP",
                   effect_col = "BETA", stderr_col = "SE", position_col = "POS", 
                   vcf_folder = vcf_folder, num_causal = 1, superpopulation = "EUR",
                   minPos=NULL, maxPos=NULL, force_new_subset = force_new_subset)  
```

## Kunkle (2019) (IGAP) {.tabset .tabset-fade .tabset-pills}

Exclude CLU-associated portion of locus by removing all SNPs over position 8:27440000.   
The CLU-gene technically starts at 8:27456253, but its associated SNPs are further upstream.
```{r AD: Kunkle (2019), results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "Kunkle_2019"

top_SNPs <- import_topSNPs(
  file_path = Directory_info(dataset_name, "topSumStats"),
  sheet = "Supplementary Table 6",
  chrom_col = "Chr", position_col = "Basepair", snp_col="SNP",
  pval_col="P", effect_col="Beta", gene_col="LOCUS",
  caption= "Kunkle (2019) (IGAP): AD GWAS Summary Stats")
# genes <- snps_to_genes(snp_list=top_SNPs$SNP )

allResults[[dataset_name]] <- finemap_geneList(top_SNPs, geneList=c("PTK2B"),   
                   file_path = Directory_info(dataset_name, "fullSumStats"),
                   # file_path="Data/Alzheimers/Kunkle_2019/Kunkle_etal_Stage1_results.txt",
                   chrom_col = "Chromosome",  position_col = "Position", pval_col = "Pvalue",
                   snp_col = "MarkerName", effect_col = "Beta", stderr_col = "SE",
                   vcf_folder = vcf_folder, num_causal = 1, superpopulation = "EUR",
                   minPos=NULL, maxPos=NULL, file_sep=" ", force_new_subset = force_new_subset,  
                   bp_distance = 500000)   #maxPos=27440000,
```

# eQTL

- AFR = AFA = African  
- AMR = Ad Mixed American  
- EAS = East Asian  
- EUR = CAU = European  
- SAS = South Asian  
- HIS = Hispanic

## MESA - AFA {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - AFA, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_AFA"
superpopulation <- "AFA"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene =  "LRRK2", 
                                         fullSS_path = Directory_info(dataset_name, "fullSumStats"), 
                                         force_new_subset = force_new_subset)
```

## MESA - CAU {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - CAU, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_CAU"
superpopulation <- "CAU"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene = "LRRK2", 
                                         fullSS_path = Directory_info(dataset_name, "fullSumStats"), 
                                         force_new_subset = force_new_subset, min_r2 = 0)
```

## MESA - HIS {.tabset .tabset-fade .tabset-pills}

Used the Ad Mixed American (AMR) reference panel given the absence of a Hispanic reference panel for 1KG_Phase1.
```{r eQTL: MESA - HIS, results = 'asis', fig.height=8, fig.width=7}
dataset_name <- "MESA_HIS"
superpopulation <- "HIS"
allResults[[dataset_name]] <- finemap_eQTL(superpopulation, gene = "LRRK2", 
                                       fullSS_path = Directory_info(dataset_name, "fullSumStats"), 
                                       force_new_subset = force_new_subset, min_r2 = 0)
```

# Merge Results

```{r Merge Results}
merged_results <- merge_finemapping_results(allResults, csv_path ="Results/merged_finemapping_results.csv")
createDT(merged_results, caption = "Finemapped Credible Sets for Multiple Datasets")
# table(as.character(merged_results$SNP))
```

