---
title: "<center><h1>Fine Mapping:</h1>Parkinson's Disease</h1></center>" 
author: 
    "<div class='container'>
     <h3>Brian M. Schilder, Bioinformatician II<br>
     Raj Lab<br>
     Department of Neuroscience<br>
     Icahn School of Medicine at Mount Sinai<br>
     NYC, New York<br>
     </h3> 
     <a href='https://github.com/RajLabMSSM/Fine_Mapping' target='_blank'><img src='./echolocatoR/images/echo_logo_sm.png'></a> 
     <a href='https://github.com/RajLabMSSM' target='_blank'><img src='./web/images/github.png'></a> 
     <a class='item' href='https://rajlabmssm.github.io/RajLab_website/' target='_blank'>
        <img src='./web/images/brain-icon.png'>
        <span class='caption'>RAJ LAB</span>
     <a href='https://icahn.mssm.edu/' target='_blank'><img src='./web/images/sinai.png'></a>
     </div>"
date: "<br>Most Recent Update:<br> `r Sys.Date()`"
output:
 html_document:
    theme: cerulean
    highlight: zenburn
    code_folding: show
    toc: true
    toc_float: true
    smooth_scroll: true
    number_sections: false
    self_contained: true
    css: ./web/css/style.css
editor_options: 
  chunk_output_type: inline
--- 

[Scripts to run CAVIAR, FINEMAP (version 1.1) and DAP-G](https://stephenslab.github.io/susie-paper/manuscript_results/pedagogical_example.html)
```{r setup, message=F, warning=F, dpi = 600}
# knitr::opts_chunk$set(error = TRUE) # Knit even with errors
# devtools::install_local("echolocatoR/", force = T)
# library("echolocatoR")
source("echolocatoR/R/MAIN.R")
# MINERVA PATHS TO SUMMARY STATISTICS DATASETS: 
Data_dirs <- read.csv("./Data/directories_table.csv")
finemap_results <- list()
coloc_results <- list()
```

  

# Nalls et. al. (2019) w/ 23andMe

## Statistical Fine-mapping {.tabset .tabset-fade .tabset-pills} 

Parkinson's Disease GWAS.
```{r PD: Nalls23andMe_2019, results = 'asis', fig.height=12, fig.width=9} 
dataset_name <- "Nalls23andMe_2019"
top_SNPs <- Nalls_top_SNPs <- import_topSNPs(
  topSS_path = Directory_info(Data_dirs, dataset_name, "topSumStats"),
  chrom_col = "CHR", position_col = "BP", snp_col="SNP",
  pval_col="P, all studies", effect_col="Beta, all studies", gene_col="Nearest Gene",
  caption= "Nalls et al. (2018) w/ 23andMe PD GWAS Summary Stats",
  group_by_locus = T, 
  locus_col = "Locus Number",
  remove_variants = "rs34637584")
# Manually add new SNP of interest
top_SNPs <-data.table::rbindlist(list(top_SNPs, 
                                      data.table::data.table(CHR=14,POS=55360203, SNP="rs3783642", 
                                                   P=0, Effect=1, Gene="ATG14", Locus=NA),
                                      data.table::data.table(CHR=12,POS=53797236, SNP="rs34656641",
                                                   P=0, Effect=1, Gene="SP1", Locus=NA),
                                      data.table::data.table(CHR=5,POS=126181862, SNP="rs959573",
                                                   P=0, Effect=1, Gene="LMNB1", Locus=NA)
                                      ))

gene_list <- unique(top_SNPs$Gene)

finemap_results[[dataset_name]] <- finemap_gene_list(top_SNPs,
                                               gene_list = "LRRK2",#unique(c("LRRK2", gene_list)),
                                               trim_gene_limits = c(T, rep(F,length(gene_list))),
                                               dataset_name = dataset_name,
                                               dataset_type = "GWAS",
                                               query_by ="coordinates",
                                               subset_path = "auto",
                                               finemap_method = c("SUSIE","FINEMAP"),
                                               #,"ABF","FINEMAP","COJO"), 
                                               force_new_LD = T,  
                                               force_new_subset = F,
                                               force_new_finemap = F,
                 # file_path = Directory_info(dataset_name, "fullSumStats"),
                 fullSS_path = "./Data/GWAS/Nalls23andMe_2019/nallsEtAl2019_allSamples_allVariants.mod.txt",
                 chrom_col = "CHR", position_col = "POS", snp_col = "RSID",
                 pval_col = "p", effect_col = "beta", stderr_col = "se", 
                 freq_col = "freq", MAF_col = "calculate",
                 A1_col = "A1",
                 A2_col = "A2",
                 bp_distance = 500000*2,
                 download_reference = T, 
                 LD_reference = "1KG_Phase1",
                 superpopulation = "EUR", 
                 LD_block = F, 
                 min_MAF = 0,
                 remove_variants = c("rs34637584"),
                 remove_correlates = c("rs34637584"),
                 conditioned_snps = "auto", # Use the lead SNP for each gene 
                 n_causal = 5, 
                 remove_tmps = F)
 
```

# Summarize Fine-mapping Results

```{r Merge Results, results = 'asis'}
merged_results <- merge_finemapping_results(finemap_results, 
                                            top_SNPs,
                                            xlsx_path ="./Data/finemapping_annotations.xlsx", 
                                            from_storage = T, 
                                            haploreg_annotation = T,
                                            biomart_annotation = T,
                                            verbose = F)
# Summarize epigenetics 
## Across all datasets and loci
epi_table <- epigenetics_summary(merged_results,
                     tissue_list = c("BRN","BLD"),
                     epigenetic_variables = c("Promoter_histone_marks","Enhancer_histone_marks"))
createDT(epi_table)

## Within LRRK2
### Enrichment 
LRRK2_dat <- data.table::fread("Data/GWAS/Nalls23andMe_2019/LRRK2/Multi-finemap/Multi-finemap_results.txt",
                        sep="\t", stringsAsFactors = F) 
epigenetics_enrichment(snp_list1 = subset(LRRK2_dat, Support >= 1)$SNP %>% unique(), 
                       snp_list2 = LRRK2_dat$SNP %>% unique(), 
                       chisq = T, 
                       fisher = T, 
                       tissue_list = c("BRN","BLD"),
                       epigenetic_variables = c("Promoter_histone_marks","Enhancer_histone_marks"))


                                
potential_missense <- SNPs_by_mutation_type(merged_results, mutation_type="missense_variant")
createDT(subset(potential_missense, consequence_type_tv=="missense_variant"))

candidate_counts <- counts_summary(top_SNPs, merged_results)
```

## LocusZoom Plots

```{r}  

```



## eQTL Boxplots: Fairfax

```{r eQTL Boxplots: Fairfax}
# snp_list <- subset(merged_results, Gene=="LRRK2")$SNP %>% unique()
snp_list <- c("rs7294519", "rs11175620", "rs76904798")

merged_eQTL_data <- eQTL_boxplots(snp_list,
             eQTL_SS_paths = file.path("Data/eQTL/Fairfax_2014",
                                       c("CD14/LRRK2/LRRK2_Fairfax_CD14.txt",
                                         "IFN/LRRK2/LRRK2_Fairfax_IFN.txt",
                                         "LPS2/LRRK2/LRRK2_Fairfax_LPS2.txt",
                                         "LPS24/LRRK2/LRRK2_Fairfax_LPS24.txt")),
             expression_paths = file.path("Data/eQTL/Fairfax_2014",
                                          c("CD14/CD14.47231.414.b.txt",
                                            "IFN/IFN.47231.367.b.txt",
                                            "LPS2/LPS2.47231.261.b.txt",
                                            "LPS24/LPS24.47231.322.b.txt")),
             genotype_path = "Data/eQTL/Fairfax_2014/geno.subset.txt",
             probe_path = "Data/eQTL/Fairfax_2014/gene.ILMN.map",
             .fam_path = "Data/eQTL/Fairfax_2014/volunteers_421.fam",
             gene = "LRRK2",
             show_plot = T,
             SS_annotations = F) 

```



## Colocalization: MESA {.tabset .tabset-fade .tabset-pills} 

```{r Colocalization: Nalls23andMe_2019 + MESA, results = 'asis', fig.height=7, fig.width=10, eval=F}  
subpops <- c("CAU","AFA","HIS")
dataset2_pathList <- paste0("./Data/eQTL/MESA_",subpops,"/LRRK2_",subpops,"_MESA.txt")

colocalize_geneList(gene_list = c("LRRK2"), 
                    dataset1_path = "./Data/GWAS/Nalls23andMe_2019/LRRK2/Multi-finemap/Multi-finemap_results.txt",
                    dataset2_pathList = dataset2_pathList,
                    dataset1_type = "cc",
                    dataset2_type = "quant",
                    shared_MAF = 1, 
                    PP_threshold = .8, 
                    save_results = T, 
                    show_plot = T)

colocalize_geneList <- function(gene_list, 
                                dataset1_path, 
                                dataset2_pathList, 
                                dataset1_type="cc",
                                dataset2_type = "quant",
                                shared_MAF=1,
                                PP_threshold=0.8,
                                save_results=T,
                                show_plot=T,
                                chrom_col = "chr", 
                                position_col = "pos_snps", 
                                snp_col="snps",
                                pval_col="pvalue", 
                                effect_col="beta", 
                                gene_col="gene_name",
                                stderr_col = "calculate",  
                                tstat_col = "statistic"){
  for(g in gene_list){
    cat('\n')
    cat("###", g, "\n") 
    for(d2 in dataset2_pathList){
      comparison_name <- paste0(basename(dataset1_path),".vs.",basename(d2))  
      subset_DT <- extract_SNP_subset(gene = gene, 
                    top_SNPs = top_SNPs, 
                    fullSS_path = fullSS_path,
                    subset_path  =  subset_path,
                    force_new_subset = force_new_subset,
                    
                    chrom_col = chrom_col, 
                    position_col = position_col, 
                    snp_col = snp_col,
                    pval_col = pval_col, 
                    effect_col = effect_col, 
                    stderr_col = stderr_col,
                    gene_col = gene_col, 
                    tstat_col = tstat_col,
                    MAF_col = MAF_col,
                    freq_col = freq_col,
                    A1_col = A1_col,
                    A2_col = A2_col,
                    
                    bp_distance = bp_distance,
                    superpopulation = superpopulation,  
                    min_POS = min_POS, 
                    max_POS = max_POS, 
                    file_sep = file_sep, 
                    query_by = query_by,
                    probe_path = probe_path
                    ) 
      
      coloc_DT <- COLOC(
          gene = g,
          dataset1_path = dataset1_path, 
          dataset2_path = d2,
          dataset1_type = dataset1_type,
          dataset2_type = dataset2_type,
          shared_MAF = shared_MAF,
          PP_threshold = PP_threshold, 
          save_results = save_results,
          show_plot = show_plot, 
          )
      coloc_DT <- cbind(Comparison = comparison_name)
      coloc_results[[dataset_name]]
    } 
  cat("\n")
  } 
  return(coloc_results %>% data.table::rbindlist())
}


```

## Colocalization: Fairfax {.tabset .tabset-fade .tabset-pills} 

```{r Colocalization: Nalls23andMe_2019 + Fairfax, results = 'asis', fig.height=7, fig.width=10, eval=T} 
# Manual check
printer("CD14")
ff <- data.table::fread("./Data/eQTL/Fairfax_2014_CD14/LRRK2/LRRK2_Fairfax_2014_CD14_subset.txt", sep="\t")
subset(ff, SNP=="rs7294619")

printer("IFN")
ff <- data.table::fread("./Data/eQTL/Fairfax_2014_IFN/LRRK2_Fairfax_IFN_mod.txt", sep="\t")
subset(ff, POS=="40617202")

printer("LPS 2-hour")
ff <- data.table::fread("./Data/eQTL/Fairfax_2014_LPS2/LRRK2/LRRK2_Fairfax_LPS2.txt", sep="\t")
subset(ff, POS=="40617202")

printer("LPS 24-hour")
ff <- data.table::fread("./Data/eQTL/Fairfax_2014_LPS24/LRRK2/LRRK2_Fairfax_LPS24.txt", sep="\t")
subset(ff, POS=="40617202")

# 
# for(g in c("LRRK2")){
#   cat('\n')
#   cat("###", g, "\n") 
#   for(s in c("IFN","LPS2", "LPS24")){ 
#     dataset_name <- paste0("Fairfax_2014_",s)
#     # Fine-map
#     try({
#     allResults[[dataset_name]] <- finemap_gene_list( gene_list=c("LRRK2"),  
#                                                 superpopulation = s,
#                                                 dataset_name = dataset_name, 
#                                                 dataset_type = "GWAS",
#                                                 finemap_method = c("FINEMAP"),#c("SUSIE","ABF","FINEMAP"),
#                                                 force_new_LD = T,
#                                                 force_new_finemap = T, 
#                                                 force_new_subset = T,  
#                                                 query_by = "probe",
#                                                 top_SNPs = Nalls_top_SNPs,  
#                    # fullSS_path = Directory_info(dataset_name, "fullSumStats"),
#                    fullSS_path = paste0("./Data/eQTL/",dataset_name,"/",g,"_Fairfax_2014_",s,"_mod.txt"),
#                    subset_path = "auto",
#                    effect_col = "beta", stderr_col = "calculate", tstat_col = "t-stat", pval_col = "p-value",
#                    n_causal = 5,  
#                    LD_block = F) 
#     })
#     # Colocalize
#     coloc_results[[dataset_name]] <- COLOC(
#         gene = g,
#         dataset1_path = paste0("./Data/GWAS/Nalls23andMe_2019/",g,"/",g,"_Nalls23andMe_2019_subset.txt"), 
#         dataset2_path = paste0("./Data/eQTL/Fairfax_2014_",s,"/",g,"/",g,"_Fairfax_2014_",s,"_subset.txt"),
#         dataset1_type = "cc",
#         dataset2_type = "quant",
#         shared_MAF = 1,
#         PP_threshold = 0.8, 
#         save_results = T,
#         show_plot=T)
#   } 
#   cat("\n")
# } 
 
 
```





# Fine-mapping eQTLs

- AFR = AFA = African  
- AMR = Ad Mixed American  
- EAS = East Asian  
- EUR = CAU = European  
- SAS = South Asian  
- HIS = Hispanic

## MESA {.tabset .tabset-fade .tabset-pills}

```{r eQTL: MESA - AFA, results = 'asis', fig.height=8, fig.width=7, eval=F}
gene_list <- c("LRRK2") 
subpops <- c("CAU","AFA","HIS")  

for(g in gene_list){
  for(s in subpops){
  dataset_name <- paste0("MESA_",s)
  finemap_results[[dataset_name]] <- finemap_gene_list( 
                                                top_SNPs = Nalls_top_SNPs, 
                                                gene_list=c("LRRK2"),  
                                                superpopulation = s, 
                                                dataset_name = dataset_name, 
                                                finemap_method = c("SUSIE"),#c("SUSIE","ABF","FINEMAP"),
                                                force_new_LD = T,
                                                force_new_finemap = T, 
                                                force_new_subset = T, 
                                                dataset_type = "eQTL",
                                                query_by = "fullSS", 
                                                file_sep = "\t",
                   # fullSS_path = Directory_info(dataset_name, "fullSumStats"),
                   fullSS_path = paste0("./Data/eQTL/MESA_",s,"/",g,"_MESA_",s,".txt"),
                   subset_path = "auto",
                   chrom_col = "chr", position_col = "pos_snps", snp_col="snps",
                   pval_col="pvalue", effect_col="beta", gene_col="gene_name", 
                   stderr_col = "calculate",  tstat_col = "statistic",
                   n_causal = 5,   
                   download_reference = T,
                   remove_tmps = F) 
  }  
}

 
```



## Fairfax (2014) {.tabset .tabset-fade .tabset-pills}

+ Use the Nalls et al. (2019) LRRK2 locus to query for SNPs in the eQTL data.  
+ SS Files are SPACE-delimited (not tab-delimited)
```{r Fairfax, results = 'asis', fig.height=8, fig.width=7, eval=F} 

# +++++++++++++++ CD14 Stimulation +++++++++++++++ #
dataset_name <- "Fairfax_2014_CD14"
allResults[[dataset_name]] <- finemap_gene_list( gene_list=c("LRRK2"),  
                                                superpopulation = "CAU",
                                                dataset_name = dataset_name, 
                                                dataset_type = "eQTL",
                                                top_SNPs = Nalls_top_SNPs, 
                                                query_by = "probes",
                                                probe_path = "./Data/eQTL/gene.ILMN.map",
                   file_sep = " ",                            
                   # fullSS_path = Directory_info(Data_dirs, dataset_name, "fullSumStats"),
                   fullSS_path = "./Data/eQTL/Fairfax_2014_CD14/LRRK2_Fairfax_CD14.txt",
                   subset_path = "auto",
                   chrom_col = "CHR", position_col = "POS", snp_col="SNP",
                   pval_col="p-value", effect_col="beta", gene_col="gene", 
                   stderr_col = "calculate",  tstat_col = "t-stat",
                   vcf_folder = vcf_folder, n_causal = 1, 
                   force_new_subset = force_new_subset,
                   LD_block = T)

```

# Session Information

```{r Session Information}
sessionInfo()
print(paste("susieR ", packageVersion("susieR")))
```


  

