<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link href="//fonts.googleapis.com/css?family=Raleway:400,300,600" rel="stylesheet" type="text/css">
    <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/skeleton/2.0.4/skeleton.css"/>
 
    <!-- Necessary includes for LocusZoom.js -->
    <script src="dist/locuszoom.vendor.min.js" type="text/javascript"></script>
    <script src="dist/locuszoom.app.min.js" type="text/javascript"></script>
    <script type="application/javascript" src="dist/ext/lz-dynamic-urls.min.js"></script>

    <link rel="stylesheet" href="dist/locuszoom.css" type="text/css"/>  
    <title>echolocatoR: LocusZoom</title>

    <style>
      body {
        background-color: #FAFAFA;
        margin: 0px 20px;
      }
      img {
        max-width: 100%;
        box-sizing: border-box;
      }
      div.example > a > h6 {
        margin-bottom: 0.5em;
      }
      div.example > a > img {
        border: 1px solid #8A8A8A;
        margin-bottom: 1.4em;
      }
    </style>

  </head>

  <body>
    <div class="container">

      <h1 style="margin-top: 1em;"><strong>LocusZoom.js</strong></h1>
      <h5 style="color: #777">Customized for use by echolocatoR</h5>

      <hr>

      <a name="try_it"></a>
     
      <div class="row">
        <div class="two columns">
          <h4>Top Hits</h4>
          <style>ul.top_hits li { margin-bottom: 0rem; }</style>
          <ul class="top_hits" style="padding-left: 0.2rem; min-width: 110px;"></ul>
        </div>
        <div class="ten columns">
          <div id="lz-plot"></div>
        </div>
      </div>
 
    <script type="text/javascript">

    // Determine if we're online, based on browser state or presence of an optional query parameter
    var online = !(typeof navigator != "undefined" && !navigator.onLine);
    if (window.location.search.indexOf("offline") != -1){ online = false; }

    // Define LocusZoom Data Sources object differently depending on online status
    var apiBase, data_sources;
    if (online){
      apiBase = "https://portaldev.sph.umich.edu/api/v1/";
      data_sources = new LocusZoom.DataSources()
        .add("assoc", ["AssociationLZ", {url: apiBase + "statistic/single/", params: { source: 45, id_field: "variant" }}])
        .add("ld", ["LDLZ2", { url: "https://portaldev.sph.umich.edu/ld/", params: { source: '1000G', build: 'GRCh37', population: 'ALL' } }])
        .add("gene", ["GeneLZ", { url: apiBase + "annotation/genes/", params: { build: 'GRCh37' } }])
        .add("recomb", ["RecombLZ", { url: apiBase + "annotation/recomb/results/", params: { build: 'GRCh37' } }])
        .add("constraint", ["GeneConstraintLZ", { url: "http://exac.broadinstitute.org/api/constraint" }]);
    } else {
      apiBase = window.location.origin + window.location.pathname.substr(0, window.location.pathname.lastIndexOf("/") + 1) + "staticdata/";
      data_sources = new LocusZoom.DataSources()
        .add("assoc", ["AssociationLZ", {url: apiBase + "assoc_10_114550452-115067678.json?", params: { source: 45, id_field: "variant" }}])
        .add("ld", ["LDLZ", { url: apiBase + "ld_10_114550452-115067678.json?" }])
        .add("gene", ["GeneLZ", { url: apiBase + "genes_10_114550452-115067678.json?" }])
        .add("recomb", ["RecombLZ", { url: apiBase + "recomb_10_114550452-115067678.json?" }])
        .add("constraint", ["GeneConstraintLZ", {  url: apiBase + "constraint_10_114550452-115067678.json?" }]);
    } 
    // Get the standard association plot layout from LocusZoom's built-in layouts
    var stateUrlMapping = {chr: "chrom", start: "start", end: "end"};
    // Fetch initial position from the URL, or use some defaults
    var initialState = LocusZoom.ext.DynamicUrls.paramsFromUrl(stateUrlMapping);
    if (!Object.keys(initialState).length) {
        initialState = {chr: 10, start: 114550452, end: 115067678};
    }
    layout = LocusZoom.Layouts.get("plot", "standard_association", {state: initialState});
    layout.dashboard = LocusZoom.Layouts.get("dashboard", "region_nav_plot");

    // Add a button to show the study abstract the layout
    var abstract = "<h4 style=\"margin-top: 0px;\"><a href=\"https://www.ncbi.nlm.nih.gov/pubmed/?term=22885922\" target=\"_new\">Type 2 diabetes meta-analysis</a></h4>1. Nat Genet. 2012 Sep;44(9):981-90. doi: 10.1038/ng.2383. Epub 2012 Aug 12.<br><br>Large-scale association analysis provides insights into the genetic architecture <br>and pathophysiology of type 2 diabetes.<br><br>Morris AP, Voight BF, Teslovich TM, Ferreira T, Segrè AV, Steinthorsdottir V,<br>Strawbridge RJ, Khan H, Grallert H, Mahajan A, Prokopenko I, Kang HM, Dina C,<br>Esko T, Fraser RM, Kanoni S, Kumar A, Lagou V, Langenberg C, Luan J, Lindgren CM,<br>Müller-Nurasyid M, Pechlivanis S, Rayner NW, Scott LJ, Wiltshire S, Yengo L,<br>Kinnunen L, Rossin EJ, Raychaudhuri S, Johnson AD, Dimas AS, Loos RJ, Vedantam S,<br>Chen H, Florez JC, Fox C, Liu CT, Rybin D, Couper DJ, Kao WH, Li M, Cornelis MC, <br>Kraft P, Sun Q, van Dam RM, Stringham HM, Chines PS, Fischer K, Fontanillas P,<br>Holmen OL, Hunt SE, Jackson AU, Kong A, Lawrence R, Meyer J, Perry JR, Platou CG,<br>Potter S, Rehnberg E, Robertson N, Sivapalaratnam S, Stančáková A, Stirrups K,<br>Thorleifsson G, Tikkanen E, Wood AR, Almgren P, Atalay M, Benediktsson R,<br>Bonnycastle LL, Burtt N, Carey J, Charpentier G, Crenshaw AT, Doney AS, Dorkhan<br>M, Edkins S, Emilsson V, Eury E, Forsen T, Gertow K, Gigante B, Grant GB, Groves <br>CJ, Guiducci C, Herder C, Hreidarsson AB, Hui J, James A, Jonsson A, Rathmann W, <br>Klopp N, Kravic J, Krjutškov K, Langford C, Leander K, Lindholm E, Lobbens S,<br>Männistö S, Mirza G, Mühleisen TW, Musk B, Parkin M, Rallidis L, Saramies J,<br>Sennblad B, Shah S, Sigurðsson G, Silveira A, Steinbach G, Thorand B, Trakalo J, <br>Veglia F, Wennauer R, Winckler W, Zabaneh D, Campbell H, van Duijn C,<br>Uitterlinden AG, Hofman A, Sijbrands E, Abecasis GR, Owen KR, Zeggini E, Trip MD,<br>Forouhi NG, Syvänen AC, Eriksson JG, Peltonen L, Nöthen MM, Balkau B, Palmer CN, <br>Lyssenko V, Tuomi T, Isomaa B, Hunter DJ, Qi L; Wellcome Trust Case Control<br>Consortium; Meta-Analyses of Glucose and Insulin-related traits Consortium<br>(MAGIC) Investigators; Genetic Investigation of ANthropometric Traits (GIANT)<br>Consortium; Asian Genetic Epidemiology Network–Type 2 Diabetes (AGEN-T2D)<br>Consortium; South Asian Type 2 Diabetes (SAT2D) Consortium, Shuldiner AR, Roden<br>M, Barroso I, Wilsgaard T, Beilby J, Hovingh K, Price JF, Wilson JF, Rauramaa R, <br>Lakka TA, Lind L, Dedoussis G, Njølstad I, Pedersen NL, Khaw KT, Wareham NJ,<br>Keinanen-Kiukaanniemi SM, Saaristo TE, Korpi-Hyövälti E, Saltevo J, Laakso M,<br>Kuusisto J, Metspalu A, Collins FS, Mohlke KL, Bergman RN, Tuomilehto J, Boehm<br>BO, Gieger C, Hveem K, Cauchi S, Froguel P, Baldassarre D, Tremoli E, Humphries<br>SE, Saleheen D, Danesh J, Ingelsson E, Ripatti S, Salomaa V, Erbel R, Jöckel KH, <br>Moebus S, Peters A, Illig T, de Faire U, Hamsten A, Morris AD, Donnelly PJ,<br>Frayling TM, Hattersley AT, Boerwinkle E, Melander O, Kathiresan S, Nilsson PM,<br>Deloukas P, Thorsteinsdottir U, Groop LC, Stefansson K, Hu F, Pankow JS, Dupuis<br>J, Meigs JB, Altshuler D, Boehnke M, McCarthy MI; DIAbetes Genetics Replication<br>And Meta-analysis (DIAGRAM) Consortium.<br><br>To extend understanding of the genetic architecture and molecular basis of type 2<br>diabetes (T2D), we conducted a meta-analysis of genetic variants on the<br>Metabochip, including 34,840 cases and 114,981 controls, overwhelmingly of<br>European descent. We identified ten previously unreported T2D susceptibility<br>loci, including two showing sex-differentiated association. Genome-wide analyses <br>of these data are consistent with a long tail of additional common variant loci<br>explaining much of the variation in susceptibility to T2D. Exploration of the<br>enlarged set of susceptibility loci implicates several processes, including<br>CREBBP-related transcription, adipocytokine signaling and cell cycle regulation, <br>in diabetes pathogenesis.<br><br>DOI: 10.1038/ng.2383 <br>PMCID: PMC3442244<br>PMID: 22885922  [PubMed - indexed for MEDLINE]</div>";
    layout.panels[0].dashboard.components.push({
        type: "menu",
        color: "yellow",
        position: "right",
        button_html: "Study Abstract",
        menu_html: abstract
    });

    // Generate the LocusZoom plot, and reflect the initial plot state in url
    window.plot = LocusZoom.populate("#lz-plot", data_sources, layout);

    // Changes in the plot can be reflected in the URL, and vice versa (eg browser back button can go back to
    //   a previously viewed region)
    LocusZoom.ext.DynamicUrls.plotUpdatesUrl(plot, stateUrlMapping);
    LocusZoom.ext.DynamicUrls.plotWatchesUrl(plot, stateUrlMapping);

    // Add a basic loader to each panel (one that shows when data is requested and hides when one rendering)
    plot.layout.panels.forEach(function(panel){
      plot.panels[panel.id].addBasicLoader();
    });

    // Create a method to parse a region string into a 600Kb genome range and load it
    function jumpTo(region) {
      var target = region.split(":");
      var chr = target[0];
      var pos = target[1];
      var start = 0;
      var end = 0;
      if (!pos.match(/[-+]/)) {
        start = +pos - 300000;
        end = +pos + 300000
      }
      plot.applyState({ chr: chr, start: start, end: end, ldrefvar: "" });
      return false;
    }

    // Populate a list of top hits links for the plot
    var top_hits = [
      ["16:53819169", "FTO"],
      ["9:22051670", "CDKN2A/B"],
      ["7:28196413", "JAZF1"],
      ["12:71433293", "TSPAN8"],
      ["10:114758349", "TCF7L2"],
      ["8:95937502", "TP53INP1"],
      ["6:20679709", "CDKAL1"],
      ["2:161346447", "RBMS1"],
      ["16:75247245", "BCAR1"],
      ["15:77832762", "HMG20A"],
      ["7:15052860", "DGKB"]
    ];
    top_hits.forEach(function(hit){
      d3.select("ul.top_hits").append("li")
        .html("<a href=\"javascript:void(0);\" onclick=\"javascript:jumpTo('" + hit[0] + "');\">" + hit[1] + "</a>");
    });  

    function read_file(file_path){
      d3.tsv(file_path, function(data){ 
        console.log(data);
        return { 
          SNP : data.SNP,
          P : +data.P,
          Effect : +data.Effect
        }  
    }) 
    }

    read_file("./Multi-finemap_results.txt"); 


     
    
 

  </script> 

  

  </body>
</html>
